{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "06cf0f38",
   "metadata": {},
   "outputs": [],
   "source": [
    "from neo4j import GraphDatabase\n",
    "from CurriculumDB.Modelsn4j import *\n",
    "import docx\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "3dd8a0f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<neo4j._sync.driver.BoltDriver object at 0x000001E81E3E0520>\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "# URI examples: \"neo4j://localhost\", \"neo4j+s://xxx.databases.neo4j.io\"\n",
    "URI = \"bolt://localhost:7687\"\n",
    "AUTH = (\"curriculum\", \"curriculumdb\")\n",
    "\n",
    "driver = GraphDatabase.driver(URI, auth=AUTH)\n",
    "print(driver)\n",
    "print(driver.verify_connectivity())\n",
    "#help(driver.verify_connectivity)\n",
    "factory = CurriculumFactory(driver, 'curriculumdb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0ef3f7fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "routes = {'BIMS': [\"BMS\",\"Biomedical Sciences\"],\n",
    " 'NEUR': [\"BMS\",\"Neuroscience\"],\n",
    " 'PHAR': [\"BMS\",\"Pharmacology\"],\n",
    " 'PHSC': [\"BMS\",\"Physiological Sciences\"],\n",
    " 'BIOLOGSCI': [\"BIO\",\"Biological Sciences\"],\n",
    " 'BIOC': [\"BIO\",\"Biochemistry\"],\n",
    " 'BSBI' : [\"BIO\",\"Biological Sciences (Bioinformatics)\"],\n",
    " 'BSPS': [\"BIO\",\"Biological Sciences (Plant Sciences)\"],\n",
    " 'MBIO': [\"BIO\",\"Microbiology\"],\n",
    " 'MOLG': [\"BIO\",\"Molecular Genetics\"],\n",
    " 'MOLB':[\"BIO\",\"Molecular Biology\"],\n",
    " 'BCDD' : [\"BIO\",\"Biological Chemistry and Drug Discovery\"]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ad1afd5a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'code': 'Programme code', 'name': 'Programme name'}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Programme.requiredParams\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1f3977e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "routeobj ={}\n",
    "for p in routes:\n",
    "    prog = factory.get_or_create_Element('Programme', code=p, name=routes[p][1])\n",
    "    routeobj[p]=prog\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7d95da15",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "doc = docx.Document(\"c:/Users/marti/Documents/LifeSciteaching/Curriculum/Programmes/UG/BSc Hons Biomedical Sciences Programme Specification 2324.docx\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "41994a31",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "class Node():\n",
    "    '''Represents a Node in a Neo4j database. \n",
    "    Contains methods for accessing links, ensuring that on creation the \n",
    "    correct parameters are given etc.'''\n",
    "\n",
    "    ntype='Node'\n",
    "    \n",
    "    requiredParams = {\n",
    "        }\n",
    "    optionalParams= {}\n",
    "    \n",
    "    def __init__(self, factory=None,  **kwparams):\n",
    "        '''Initialiser. Checks required parameters are present'''\n",
    "        if factory is None:\n",
    "            raise Exception('Curriculum factory must be given as factory')\n",
    "        self.factory = factory\n",
    "        self.ntype=self.__class__.__name__.split('.')[-1]\n",
    "        self.paramChecks={'default': self._checkDefault}\n",
    "        self.params={}\n",
    "        self.edges=[]\n",
    "        self._get_labels()\n",
    "        for par in self.requiredParams:\n",
    "            if not kwparams.get(par):\n",
    "                print(par, kwparams)\n",
    "                raise InsufficientParameterSpecException(f'Missing parameter <{par}> of type <{self.requiredParams[par]}>')\n",
    "            if not self.paramChecks.get(par,self._checkDefault)(kwparams[par]):\n",
    "                raise BadParameterException(f'Parameter <{par}> should be of type <{self.requiredParams[par]}>')\n",
    "            self.params[par]=kwparams[par]\n",
    "        for par in self.optionalParams:\n",
    "            if par not in kwparams:\n",
    "                continue\n",
    "            if not self.paramChecks.get(par,self._checkDefault)(kwparams[par]):\n",
    "                raise BadParameterException(f'Parameter <{par}> should be of type <{self.requiredParams[par]}>')\n",
    "            self.params[par]=kwparams[par]\n",
    "        if 'elementID' not in kwparams:\n",
    "            self._create_node()\n",
    "        else:\n",
    "            self.element_id=kwparams['elementID']\n",
    "            self.edges =self.getEdges()\n",
    "    def __str__(self):\n",
    "        '''Returns a string represntation of the node'''\n",
    "        params = \", \".join([f'{k}: {self.params[k]}' for k in self.requiredParams])\n",
    "        return f'{self.ntype}: [{params}] ID: {self.element_id}'\n",
    "    \n",
    "    def __repr__(self):\n",
    "        return self.__str__()\n",
    "                            \n",
    "    def _create_node(self):\n",
    "        '''Internal method that creates a new instance of the node, if it does not exist'''\n",
    "        paramlist = \", \".join([f'{k}: ${k}' for k in self.params])\n",
    "        paramlist = \"{\"+paramlist+\"}\"\n",
    "        nodelabels = ':'.join(self.labels)\n",
    "        query= f\"MERGE (a:{nodelabels} {paramlist}) RETURN a\"\n",
    "        records,summary, keys = self.factory.db.execute_query(query, self.params, database_=self.factory.dbname)   \n",
    "        self.element_id = records[0].items()[0][1].element_id\n",
    "            \n",
    "    def _checkDefault(self, value):\n",
    "        '''default placeholder for parameter value checks. Returns True'''\n",
    "        return True\n",
    "    \n",
    "    def _checkAY(self, value):\n",
    "        '''Checks for a correctly formatted Academic year as:\n",
    "            2425\n",
    "            24/25\n",
    "            2024/5'''\n",
    "        #TODO \n",
    "        return True\n",
    "    def getparam(self, prop, default=None):\n",
    "        '''\n",
    "        Returns a parameter value (default if not found)\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        prop : property name\n",
    "            \n",
    "        default : TYPE, optional\n",
    "             The default is None.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        Property value\n",
    "            \n",
    "        '''\n",
    "        return self.params.get(prop,default)\n",
    "    \n",
    "    def setparam(self, key, value):\n",
    "        '''\n",
    "        Sets the parameter key to value. No type checking.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        key : legal value for a dictionary key\n",
    "            \n",
    "        value : Value to store.\n",
    "        \n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        self.params[key]=value\n",
    "        self.update()\n",
    "    \n",
    "    def getEdges(self,relation=None,**kwargs):\n",
    "        '''retrieves edges connected to or from the Node.\n",
    "        Relation is the type of relation to retrieve (default is all)\n",
    "        other arguments are applied as matching parameters'''\n",
    "        edges=[]\n",
    "        filterstring=''\n",
    "        filterparams={}\n",
    "        for kw in kwargs:\n",
    "            if kw != 'id':\n",
    "                filterstring += f\"'{kw}': ${kw},\"\n",
    "                filterparams[kw] = kwargs[kw]\n",
    "        rel=''\n",
    "        if relation:\n",
    "            rel=f\":{relation}\"\n",
    "        filterparams['id'] = self.element_id\n",
    "        if filterstring:\n",
    "            filterstring = \"{\"+filterstring+\"}\"\n",
    "        records, summary, keys = self.factory.db.execute_query(\n",
    "        f\"MATCH (p: {self.ntype} {{id: $id }}) -[e{rel} {filterstring}]-(q) RETURN e,q\",\n",
    "        filterparams,\n",
    "        routing_=neo4j.RoutingControl.READ,  # or just \"r\"\n",
    "        database_=self.factory.dbname)\n",
    "        for edge in records:\n",
    "            relation = edge.items[0][1]\n",
    "            target = edge.items[1][1]\n",
    "            edges.append({'source': self,'edge': relation, 'target': target})\n",
    "        return edges\n",
    "    def isCurrent(self, year):\n",
    "        '''\n",
    "        If the edge has a start/end year, check if the given year is the period in which the relation is valid\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        Boolean: True if the eyar is between start and end (inclusive), after start if no end, or True if no start\n",
    "\n",
    "        '''\n",
    "        \n",
    "        if self.params.get('startyear'):\n",
    "            if AcademicYear(year)< AcademicYear(self.params.get('startyear')):\n",
    "                return False\n",
    "            if self.params.get('endyear'):\n",
    "                if AcademicYear(year)> AcademicYear(self.params.get('endyear')):\n",
    "                    return False\n",
    "        return True\n",
    "    \n",
    "    \n",
    "    def toDict(self):\n",
    "        retp ={'element_id': self.element_id}\n",
    "        for p in self.params:\n",
    "            retp[p]=self.params[p]\n",
    "        return retp\n",
    "    \n",
    "    def update(self, **kwargs):\n",
    "        '''\n",
    "        Update named parameters for a Node\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        **kwargs : key,value parameter set\n",
    "            any key,value parameters. Existing values will be overwritten.\n",
    "            No type checking\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        paramlist =[]\n",
    "        if not kwargs:\n",
    "            return\n",
    "        for k in kwargs:\n",
    "            if k !='id':\n",
    "                paramlist.append( f'a.{k}= ${k}')\n",
    "                self.params[k]=kwargs[k]\n",
    "        kwargs['id'] = self.element_id    \n",
    "        paramlist = \"{\"+paramlist+\"}\"\n",
    "        query= f\"MERGE (a:{self.ntype}) WHERE elementID(a) = $id SET {', '.join(paramlist)}\"\n",
    "        records,summary, keys = self.factory.db.execute_query(query,kwargs, database_=self.factory.dbname)   \n",
    "        \n",
    "    def _get_labels(self):\n",
    "        self.labels = [x.__name__ for x in globals()[self.__class__.__name__].__mro__]\n",
    "        self.labels.remove('object')\n",
    "        self.labels.remove('Node')\n",
    "        \n",
    "    \n",
    "    \n",
    "class Edge():\n",
    "    '''Represents an edge in a Neo4J database'''\n",
    "    requiredParameters={\n",
    "        'relation': 'relationship type',\n",
    "        'source': 'Node ',\n",
    "        'target': 'Node '}\n",
    "    optionalParams={\n",
    "        'startyear': 'Academic year start',\n",
    "        'endyear': 'Academic year to'\n",
    "        }\n",
    "    def __init__(self, factory, **kwargs):\n",
    "        '''Edge between two nodes. Must specify source, target and relationship type.'''\n",
    "        for kw in self.requiredParameters:\n",
    "            if kw not in kwargs:\n",
    "                raise InsufficientParameterSpecException(f\"{kw} not specified in Edge constructor\")\n",
    "            self.params[kw]=kwargs[kw]\n",
    "    def set_expiry(self, expirydate):\n",
    "        '''\n",
    "        Sets the 'to' parameter with an academic year for when this relation was withdrawn. \n",
    "        The date specified will be the last academic year for which it was valid.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        expirydate : Academic year\n",
    "            Academic year is an academic year object or text description of it.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        self.params['to'] =AcademicYear(expirydate)\n",
    "    \n",
    "    def isCurrent(self, year):\n",
    "        '''\n",
    "        If the edge has a start/end year, check if the given year is the period in which the relation is valid\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        Boolean: True if the eyar is between start and end (inclusive), after start if no end, or True if no start\n",
    "\n",
    "        '''\n",
    "        \n",
    "        if self.params.get('startyear'):\n",
    "            if AcademicYear(year)< AcademicYear(self.params.get('startyear')):\n",
    "                return False\n",
    "            if self.params.get('endyear'):\n",
    "                if AcademicYear(year)> AcademicYear(self.params.get('endyear')):\n",
    "                    return False\n",
    "        return True\n",
    "    def getSource(self):\n",
    "        '''\n",
    "        Returns a Python object representing the source node\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        \n",
    "class CurriculumFactory():\n",
    "    \n",
    "    def __init__(self, db, dbname='curriculum'):\n",
    "        self.db = db\n",
    "        self.dbname = dbname\n",
    "        self.programmecache = {}\n",
    "        self.modulecache = {}\n",
    "        self.activitycache = {}\n",
    "        self.personcache = {}\n",
    "        \n",
    "    def get_connection(self):\n",
    "        return self.db\n",
    "    \n",
    "    @lru_cache    \n",
    "    def get_element_by_ID(self, elementID):\n",
    "        '''\n",
    "        Retrieves an element by ID and creates the appropriate node for it, if it is a node.\n",
    "        \n",
    "        Parameters\n",
    "        ----------\n",
    "        elementID : Internal neo4j element ID\n",
    "            text string uniquely identifying the object in the database.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        Object of correct class, if available or None if not found.\n",
    "\n",
    "        '''\n",
    "        erecords, _, _ = self.db.execute_query(\n",
    "            \"MATCH (a ) where elementId(a)=$id\"\n",
    "            \"RETURN a\", id=elementID,\n",
    "             database_=self.dbname,\n",
    "        )\n",
    "        if not erecords :\n",
    "            return\n",
    "        nodeclass = list(erecords[0].items()[0][1].labels())[0]\n",
    "        params = dict(erecords[0].items()[0][1])\n",
    "        params['elementID']=elementID\n",
    "        if nodeclass in globals():\n",
    "            return globals()[nodeclass](**params)\n",
    "        \n",
    "      \n",
    "    def get_or_create_Element(self, ElementName,**kwargs):\n",
    "        '''\n",
    "        Generic get and create method. This requires all elements to have a unique id under the property code\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        ElementName : Name of element to create, eg Programme\n",
    "            This will directly map to the class name.\n",
    "        **kwargs : Parameters to use. These will be checked against every class.\n",
    "            \n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        Element of type ElementName on success\n",
    "\n",
    "        '''\n",
    "        \n",
    "        elem= None\n",
    "        \n",
    "        if ElementName not in globals():\n",
    "            return\n",
    "        ElementClass=globals()[ElementName]\n",
    "        if kwargs.get('elementID',None):\n",
    "            elem = self.get_element_by_id(kwargs['elementID'])\n",
    "            return elem\n",
    "        if kwargs.get('code'):\n",
    "            try:\n",
    "                result,_,_=self.db.execute_query(f'MATCH (p:{ElementName} {{code: $name}} ) return p', name=kwargs['code'], database_=self.dbname)\n",
    "                if result:\n",
    "                    params=dict(result[0].items()[0][1])\n",
    "                    params['elementID']=result[0]['p'].element_id\n",
    "                    elem = ElementClass(self, **params)\n",
    "                    return elem\n",
    "            except Exception as e:\n",
    "                e.add_note(str(params))\n",
    "                raise e\n",
    "                \n",
    "        for p in ElementClass.requiredParams:\n",
    "            if p not in kwargs:\n",
    "                raise InsufficientParameterSpecException(f'Not enough parameters specified for {ElementName}: missing{p}')\n",
    "        \n",
    "        elem = ElementClass(self, **kwargs)\n",
    "        return elem\n",
    "\n",
    "\n",
    "    def get_all_elements(self, ElementName, **params):\n",
    "        '''\n",
    "        Returns all elements of type ElementName\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        ElementName : name of an element\n",
    "            This can be any element type for which a model exists.\n",
    "        **params : Keyword parameters to use in the search\n",
    "            Exact matches only at present.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        List of elements\n",
    "\n",
    "        '''\n",
    "        \n",
    "        elements = []\n",
    "        if ElementName not in globals():\n",
    "            return elements\n",
    "        paramdetails =[]\n",
    "        paramtext=''\n",
    "        for p in params:\n",
    "            paramdetails.append(f'{p}: ${p}')\n",
    "        if paramdetails:\n",
    "            paramtext= f'{ {\",\".join(paramdetails)} }'\n",
    "        erecords, _, _ = self.db.execute_query(\n",
    "            f\"MATCH (a :{ElementName} {paramtext}) RETURN a\",\n",
    "            params,\n",
    "             database_=self.dbname,\n",
    "        )\n",
    "        try:\n",
    "            if not erecords :\n",
    "                return elements\n",
    "            for item in erecords:\n",
    "                elem=item.items()[0][1]\n",
    "                nodeclass = list(elem.labels)[0]\n",
    "                params = dict(elem)\n",
    "                params['elementID']=elem.element_id\n",
    "                if nodeclass in globals():\n",
    "                    elements.append( globals()[nodeclass](self, **params))\n",
    "            return elements  \n",
    "        except Exception as e:\n",
    "            print(e)\n",
    "            return erecords\n",
    "    def getElementsForElement(self, element, target, max_steps=2,relation=None):\n",
    "        '''\n",
    "        Retreive all elements of type Target linked to element,optionally by relation (or all relations) \n",
    "        at a maximum distance of max_steps (default 2)\n",
    "        \n",
    "        Parameters\n",
    "        ----------\n",
    "        element : TYPE\n",
    "            DESCRIPTION.\n",
    "        target : TYPE\n",
    "            DESCRIPTION.\n",
    "        max_steps: integer\n",
    "            Maximum link number to explore (default 2)\n",
    "        relation: text\n",
    "            Limit relations to those of type relation\n",
    "        Returns\n",
    "        -------\n",
    "        List of Node type objects.\n",
    "        '''\n",
    "        if target not in globals():\n",
    "            return []\n",
    "        cypher = f\"MATCH (e) --{{1,{max_steps}}} (thing:$target) where elementID(e)=$id RETURN DISTINCT thing\"\n",
    "        result,_,_=self.db.execute_query(cypher, target=target, id=element.element_id,database_=self.dbname)\n",
    "        elems = []\n",
    "        if result:\n",
    "            for elem in result:\n",
    "                item = elem.items()[0][1]\n",
    "                params =dict(item)\n",
    "                params['elementID'] = item.element_id\n",
    "                elems.append(globals()[target](self, **params))\n",
    "        return elems\n",
    "         \n",
    "class Programme(Node):\n",
    "    \n",
    "    DRAFT = 0\n",
    "    CURRENT = 1\n",
    "    ARCHIVED = 2\n",
    "    WITHDRAWN =3\n",
    "    ntype='Programme'\n",
    "    requiredParams = {\n",
    "        'code': 'Programme code',\n",
    "        'name': 'Programme name'\n",
    "        }\n",
    "    optionalParams= {\n",
    "        'startyear':'First academic year',\n",
    "        'endyear': 'Last academic year',\n",
    "        'school':'School which manages the Programme'}\n",
    "    def __init__(self, factory, **params ):\n",
    "        '''\n",
    "        Create a Programme instance.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        factory - database connection and cache for Programme\n",
    "        **params : \n",
    "            Create or replace table Programme (\n",
    "        '''\n",
    "        super().__init__(factory,**params)\n",
    "        \n",
    "        self.modules = {}\n",
    "        self.ILO = []\n",
    "        if self.element_id:\n",
    "            self.loadmodules()\n",
    "            self.loadILO()\n",
    "        \n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    def loadmodules(self):\n",
    "        '''\n",
    "        Loads module links from the database to the modules list. Does not create now entries in the database.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        self.modules={}\n",
    "        cypher = \"MATCH (p:Programme ) -[a]-(b:Module) WHERE elementID(p) = $id RETURN a,b\"\n",
    "        records,_,_ =self.factory.db.execute_query(cypher, id=self.element_id, database_=self.factory.dbname)\n",
    "        for t in records:\n",
    "            relation = t.items()[0][1]\n",
    "            target = dict(t.items()[1][1])\n",
    "            target['element_id'] = t.items()[1][1].element_id\n",
    "            if target['code'] not in self.modules:\n",
    "                self.modules[target['code']]=[]\n",
    "            self.modules[target['code']].append({'relation': {'type': relation.type, 'params':dict(relation)}, \n",
    "                                                        'target':{ 'moduleID': target['element_id'],'params':target}})\n",
    "            \n",
    "        \n",
    "\n",
    "        \n",
    "    def loadILO(self):\n",
    "        '''\n",
    "        Loads ILO links from the database. Does not create new links or ILOs.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        self.ILO = {}\n",
    "        cypher = \"MATCH (p:Programme ) -[a]-(b:ProgrammeILO) WHERE elementID(p) = $id RETURN a,b\"\n",
    "        records,_,_ =self.factory.db.execute_query(cypher, id=self.element_id, database_=self.factory.dbname)\n",
    "        for t in records:\n",
    "            relation, target = t.items()[0:2]\n",
    "            \n",
    "            self.ILO[target.element_id]= (dict(target), dict(relation))\n",
    "        \n",
    " \n",
    "        \n",
    "    def withdraw(self, year):\n",
    "        '''\n",
    "        Set the Programme to WITHDRAWN. Only possible if there are no future versions.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        if AcademicYear(year) and AcademicYear(year) > AcademicYear(self.params['startyear']):\n",
    "            self.params['endyear']=year\n",
    "            cypher = 'MATCH (p:Programme) where elementID(p)=$id SET p.endyear=$year'\n",
    "            records,_,_ =self.factory.db.execute_query(cypher, id=self.element_id,year=year, database_=self.factory.dbname)\n",
    "        \n",
    "\n",
    "    def map_module(self, module, optional=0, year=None, remove=False):\n",
    "        '''\n",
    "        Add a module to the Programme, updating if exisitng  already appended.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        module : Module object\n",
    "            DESCRIPTION.\n",
    "        optional: Boolean\n",
    "        remove: Academic Year for last instance\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        relations=('IS_CORE','IS_ELECTIVE')\n",
    "        relation= relations[int(bool(optional))]\n",
    "        cypher1 = 'MATCH (p:Programme) where elementID(p)=$pid \\n MATCH (m:Module) where elementID(m)=$mid \\n merge (p) <-[b:{relation}]-(m)  return p,b,m'\n",
    "        \n",
    "        if module.params['code'] in self.modules:\n",
    "            if remove:\n",
    "                \n",
    "                records,_,_ =self.factory.db.execute_query(cypher1.format(relation=relation), \n",
    "                                                           pid=self.element_id,year=year, \n",
    "                                                           mid=module.element_id,\n",
    "                                                           database_=self.factory.dbname)\n",
    "                for r in self.modules[module.params['code']]:\n",
    "                    if r['relation']['type']==relation:\n",
    "                        r['relation']['params']['endyear']=year\n",
    "            else:\n",
    "                for r in self.modules[module.params['code']]:\n",
    "                    if r['relation']['type']!=relation:\n",
    "                        r['relation']['params']['endyear']=year\n",
    "                        records,_,_ =self.factory.db.execute_query(cypher1.format(relation=relations[(int(bool(optional))+1)%2]), \n",
    "                                                                   pid=self.element_id,year=year, \n",
    "                                                                   mid=module.element_id,\n",
    "                                                                   database_=self.factory.dbname)\n",
    "         \n",
    "        if not remove:\n",
    "            if module.params['code'] not in self.modules:\n",
    "                 self.modules[module.params['code']]=[]\n",
    "            cypher2 =f'MATCH (p:Programme) where elementID(p)=$pid MATCH (m:Module)  where elementID(m)=$mid MERGE (p)<-[:{relation} {{startyear:$year }}] - (m) return m,p'\n",
    "            records,_,_ =self.factory.db.execute_query(cypher2, \n",
    "                                                        pid=self.element_id,year=year, \n",
    "                                                        mid=module.element_id,\n",
    "                                                        database_=self.factory.dbname)\n",
    "        self.loadmodules()\n",
    "           \n",
    "                                           \n",
    "               \n",
    "           \n",
    "           \n",
    "    def map_ilo(self, ilo, year, remove=False):\n",
    "        '''\n",
    "        Associates or updates a Programme ILO mapping to a Programme.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        ilo : ProgrammeILO \n",
    "            Programme ILO object\n",
    "        year : text describing the academic year\n",
    "            Must be either the start year, or for removing an ILO the last year for which it will be relevant.\n",
    "        remove : Boolean, optional\n",
    "            Flag to say whether the ILO should be dissociated from teh Programme. The default is False.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        \n",
    "        if not AcademicYear(year):\n",
    "            raise UnparseableYearException(f'Cannot parse year value {year}')\n",
    "        if ilo.element_id in self.ILO:\n",
    "            if remove:\n",
    "                self.ILO[ilo.element_id]['endyear']=year\n",
    "                cypher=\"MATCH (p: Programme) -[b]- (i:ProgrammeILO) WHERE elementID(p) =$pid AND elementID(i) =$iid SET b.endyear=$year\"\n",
    "                records,_,_ = self.factory.db.execute_query(cypher,pid=self.element_id, iid=ilo.element_id, year=year, database_=self.factory.dbname)\n",
    "            return\n",
    "        if remove:\n",
    "            return\n",
    "        cypher = \"MERGE (p:Programme) -[b:HAS_ILO {startyear:$year}]->(i:ProgrammeILO) where elementID(p)=$pid AND elementID(i) = $iid RETURN b\"\n",
    "        records,_,_ = self.factory.db.execute_query(cypher,pid=self.element_id, iid=ilo.element_id, year=year, database_=self.factory.dbname)\n",
    "        if records:\n",
    "            relation=records[0].items()[0][1]\n",
    "            self.ILO[ilo.element_id]=(dict(ilo),dict(relation))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "7d48ef50",
   "metadata": {},
   "outputs": [],
   "source": [
    "factory = CurriculumFactory(driver, 'curriculumdb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "55b0e819",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Core', 'BS11009', 'The Building Blocks of Life', '1', '7', '20', 'SLS']\n",
      "['Core', 'BS11008', 'Core Skills in the Life Sciences 1A', '1', '7', '20', 'SLS']\n",
      "['Core', 'BS12011', 'Building the Organism', '1', '7', '20', 'SLS']\n",
      "['Core', 'BS12010', 'Core Skills in the Life Sciences 1B', '1', '7', '20', 'SLS']\n",
      "['Core', 'BS21001', 'Statistics and Experimental Design', '2', '8', '10', 'SLS']\n",
      "['Core', 'BS21002', 'Cellular Communication', '2', '8', '10', 'SLS']\n",
      "['Core', 'BS21012', 'Core Skills in the Life Sciences 2A', '2', '8', '20', 'SLS']\n",
      "['Core', 'BS22001', 'Human Physiology and Pharmacology', '2', '8', '20', 'SLS']\n",
      "['Core', 'BS22002', 'Biomolecular Mechanisms', '2', '8', '20', 'SLS']\n",
      "['Core', 'BS22003', ' Core Skills in the Life Sciences 2B', '2', '8', '20', 'SLS']\n",
      "['Core', 'BS31013', 'Biomembranes', '3', '9', '15', 'SLS/SoM']\n",
      "['Core', 'BS31016', 'Practical Techniques in Biomedical Sciences', '3', '9', '15', 'SLS/SoM']\n",
      "['Core', 'BS31019', 'Regulatory Physiology & Pharmacology', '3', '9', '15', 'SLS/SoM']\n",
      "['Core', 'BS30007**', 'Life Sciences Placement', '3', '9', '80', 'SLS ']\n",
      "['Core', 'BS41006', 'Research Skills – Biomedical Sciences: Biological Sciences', '4', '10', '20', 'SLS/SoM']\n",
      "['Core', 'BS41007', 'Research project – Biomedical Sciences', '4', '10', '40', 'SLS/SoM']\n",
      "['Elective', 'BS11005', 'Numeracy, Chemistry and Physics for the Biological and Biomedical Sciences', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS11006', 'The Poison Pen', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS12005', 'Science and Society', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS12008', 'Introduction to Scientific Enterprise', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS12009', 'Fundamentals of Organic Chemistry', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS21008', 'Organic Chemistry for the Life Sciences', '2', '8', '20', 'SLS']\n",
      "['Elective', 'BS21009', 'Data Analysis for the Life Sciences', '2', '8', '20', 'SLS']\n",
      "['Elective', 'BS21010', 'Introductory Programming for the Life Sciences', '2', '8', '20', 'SLS']\n",
      "['Elective', 'BS21011', 'Developing Scientific Enterprise', '2', '8', '20', 'SLS']\n",
      "['Elective', 'BS31004', 'Biochemistry and Cell Biology', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS31006', 'Gene Regulation and Expression', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS31020', 'Experimental Cell Culture ', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32006', 'Cell Signalling*', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32026', 'Sensory & Motor Neuroscience', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32028', 'Molecular Pharmacology', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS32029', 'Cell Proliferation and Survival Mechanisms Underlying Disease', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS32030', 'Data and Statistical Analysis*', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32031', 'Bioinformatics Research Skills I*', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32033', 'Epidemiology and Health Informatics', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS32005', 'Cell and Developmental Biology*', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32009', 'Immunology*', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32020', 'Human Epithelial Biology', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32021', 'Quantitative Pharmacology', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32024', 'Behavioural Neuroscience and Pharmacology', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS32032', 'Science Communication', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42003', 'Advanced Bioinformatics', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS42006', 'Advanced Immunology*', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS42007', 'Cancer Biology', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42008', 'Stem Cells in Development and Disease*', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS42009', 'Advanced Molecular Microbiology*', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS42012', 'Parasitology*', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS42013', 'Advanced Cell Signalling*', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS42014', 'Nutrients & Metabolic Disease', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42017', 'Translational Psychiatry & Neurology', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42018', 'Synaptic Plasticity & Cognition', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42019', 'Cardiovascular Pharmacology', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42021', 'Heart & Circulation', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42023', 'Applied Neuroanatomy & Neurodegenerative Disorders', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42025', 'The Conscious Brain: The Physiology & Pharmacology of Pain and Anaesthesia', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42027', 'Cancer Pharmacology & Treatment', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42028', 'Pharmacology and Treatment of Metabolic Disease ', '4', '10', '15', 'SLS/SoM']\n"
     ]
    }
   ],
   "source": [
    "modmap = []\n",
    "structure ={}\n",
    "for t in doc.tables:\n",
    "    for r in range(len(t.rows)):\n",
    "        #print(len(t.rows), len(t.row_cells(r)), t.row_cells(r)[0].text,[c.tables for c in t.row_cells(r)])\n",
    "        structure[t.row_cells(r)[0].text.split()[0]] = t.row_cells(r)\n",
    "tc=0\n",
    "for m in structure['2.10'][0].tables:\n",
    "    coretype=['Core','Elective'][tc%2]\n",
    "    tc+=1\n",
    "    for r in range(1,len(m.rows)):\n",
    "        print([coretype]+[c.text for c in m.row_cells(r)])\n",
    "        modmap.append([coretype]+[c.text for c in m.row_cells(r)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "f1be9a3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "mods = factory.get_all_elements('Module')\n",
    "modules ={m.params['code']:m for m in mods} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "886a0da7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BS11005': Module: [code: BS11005, name: Numeracy, Chemistry and Physics for the Biological and Biomedical Sciences, credits: 20, scqflevel: 7, shelevel: 1, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:12,\n",
       " 'BS11006': Module: [code: BS11006, name: The Poison Pen, credits: 20, scqflevel: 7, shelevel: 1, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:13,\n",
       " 'BS11008': Module: [code: BS11008, name: Core Skills in the Life Sciences 1A, credits: 20, scqflevel: 7, shelevel: 1, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:14,\n",
       " 'BS11009': Module: [code: BS11009, name: The Building Blocks of Life, credits: 20, scqflevel: 7, shelevel: 1, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:15,\n",
       " 'BS12005': Module: [code: BS12005, name: Science and Society, credits: 20, scqflevel: 7, shelevel: 1, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:16,\n",
       " 'BS12008': Module: [code: BS12008, name: Introduction to Scientific Enterprise, credits: 20, scqflevel: 7, shelevel: 1, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:17,\n",
       " 'BS12009': Module: [code: BS12009, name: Fundamentals of Organic Chemistry, credits: 20, scqflevel: 7, shelevel: 1, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:18,\n",
       " 'BS12010': Module: [code: BS12010, name: Core Skills in the Life Sciences 1B, credits: 20, scqflevel: 7, shelevel: 1, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:19,\n",
       " 'BS12011': Module: [code: BS12011, name: Building the Organism, credits: 20, scqflevel: 7, shelevel: 1, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:20,\n",
       " 'BS12701': Module: [code: BS12701, name: GES1xxx: Singapore Studies, credits: 12, scqflevel: 7, shelevel: 1, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:21,\n",
       " 'BS12702': Module: [code: BS12702, name: GEXxxxx: General Education, credits: 12, scqflevel: 7, shelevel: 1, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:22,\n",
       " 'BS21001': Module: [code: BS21001, name: Statistics and Experimental Design, credits: 10, scqflevel: 8, shelevel: 2, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:23,\n",
       " 'BS21002': Module: [code: BS21002, name: Cellular Communication, credits: 10, scqflevel: 8, shelevel: 2, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:24,\n",
       " 'BS21008': Module: [code: BS21008, name: Organic Chemistry for the Life Sciences, credits: 20, scqflevel: 8, shelevel: 2, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:25,\n",
       " 'BS21010': Module: [code: BS21010, name: Introductory programming for Life Sciences, credits: 20, scqflevel: 8, shelevel: 2, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:26,\n",
       " 'BS21011': Module: [code: BS21011, name: Developing Scientific Enterprise, credits: 20, scqflevel: 8, shelevel: 2, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:27,\n",
       " 'BS21012': Module: [code: BS21012, name: Core Skills in the Life Sciences 2A, credits: 20, scqflevel: 8, shelevel: 2, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:28,\n",
       " 'BS22001': Module: [code: BS22001, name: Human Physiology and Pharmacology, credits: 20, scqflevel: 8, shelevel: 2, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:29,\n",
       " 'BS22002': Module: [code: BS22002, name: Biomolecular Mechanisms, credits: 20, scqflevel: 8, shelevel: 2, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:30,\n",
       " 'BS22003': Module: [code: BS22003, name: Core Skills in the Life Sciences 2B, credits: 20, scqflevel: 8, shelevel: 2, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:31,\n",
       " 'BS31003': Module: [code: BS31003, name: Molecular Structure & Interactions, credits: 15, scqflevel: 9, shelevel: 3, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:32,\n",
       " 'BS31004': Module: [code: BS31004, name: Biochemistry and Cell Biology, credits: 15, scqflevel: 9, shelevel: 3, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:33,\n",
       " 'BS31005': Module: [code: BS31005, name: Genetics, credits: 15, scqflevel: 9, shelevel: 3, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:34,\n",
       " 'BS31006': Module: [code: BS31006, name: Gene Regulation & Expression, credits: 15, scqflevel: 9, shelevel: 3, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:35,\n",
       " 'BS31013': Module: [code: BS31013, name: Biomembranes , credits: 15, scqflevel: 9, shelevel: 3, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:36,\n",
       " 'BS31016': Module: [code: BS31016, name: Practical Techniques in Biomedical Sciences, credits: 15, scqflevel: 9, shelevel: 3, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:37,\n",
       " 'BS31019': Module: [code: BS31019, name: Regulatory Physiology & Pharmacology, credits: 15, scqflevel: 9, shelevel: 3, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:38,\n",
       " 'BS31020': Module: [code: BS31020, name: Experimental Cell Culture, credits: 15, scqflevel: 9, shelevel: 3, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:39,\n",
       " 'BS31701': Module: [code: BS31701, name: LSM3217: Human Ageing, credits: 12, scqflevel: 10, shelevel: 4, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:40,\n",
       " 'BS31702': Module: [code: BS31702, name: LSM3222: Human Neuroanatomy, credits: 12, scqflevel: 9, shelevel: 3, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:41,\n",
       " 'BS32003': Module: [code: BS32003, name: Drug Discovery and Development, credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:42,\n",
       " 'BS32004': Module: [code: BS32004, name: Molecular Microbiology, credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:43,\n",
       " 'BS32005': Module: [code: BS32005, name: Developmental Biology, credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:44,\n",
       " 'BS32006': Module: [code: BS32006, name: Cell Signalling, credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:45,\n",
       " 'BS32007': Module: [code: BS32007, name: Organic Synthesis, credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:46,\n",
       " 'BS32008': Module: [code: BS32008, name: Plant Science, credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:47,\n",
       " 'BS32009': Module: [code: BS32009, name: Immunology, credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:48,\n",
       " 'BS32010': Module: [code: BS32010, name: Applied Bioinformatics, credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:49,\n",
       " 'BS32011': Module: [code: BS32011, name: Practical Project A, credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:50,\n",
       " 'BS32012': Module: [code: BS32012, name: Practical Project B, credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:51,\n",
       " 'BS32020': Module: [code: BS32020, name: Human Epithelial Biology, credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:52,\n",
       " 'BS32021': Module: [code: BS32021, name: Quantitative Pharmacology, credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:53,\n",
       " 'BS32024': Module: [code: BS32024, name: Behavioural Neuroscience and Pharmacology, credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:54,\n",
       " 'BS32026': Module: [code: BS32026, name: Sensory and Motor Neuroscience, credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:55,\n",
       " 'BS32028': Module: [code: BS32028, name: Molecular Pharmacology, credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:56,\n",
       " 'BS32029': Module: [code: BS32029, name: Cell proliferation and survival mechanisms underlying disease, credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:57,\n",
       " 'BS32030': Module: [code: BS32030, name: Data and Statistical Analysis, credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:58,\n",
       " 'BS32031': Module: [code: BS32031, name: Bioinformatics Research Skills 1, credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:59,\n",
       " 'BS32032': Module: [code: BS32032, name: Science Communication, credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:60,\n",
       " 'BS32033': Module: [code: BS32033, name: Epidemiology and Health Informatics, credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:61,\n",
       " 'BS32034': Module: [code: BS32034, name: Introduction to Biobusiness , credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:62,\n",
       " 'BS32035': Module: [code: BS32035, name: Project Management and Bioenterprise, credits: 15, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:63,\n",
       " 'BS32701': Module: [code: BS32701, name: LSM3211: Fundamental Pharmacology, credits: 12, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:64,\n",
       " 'BS32702': Module: [code: BS32702, name: LSM3212: Human Physiology: Cardiopulmonary System, credits: 12, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:65,\n",
       " 'BS32703': Module: [code: BS32703, name: LSM3214: Human Physiology - Hormones And Health, credits: 12, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:66,\n",
       " 'BS32704': Module: [code: BS32704, name: LSM3216: Neuronal Development and Diseases, credits: 12, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:67,\n",
       " 'BS32705': Module: [code: BS32705, name: LSM3219: Neuropharmacology, credits: 12, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:68,\n",
       " 'BS32706': Module: [code: BS32706, name: LSM3224: Molecular Basis of Human Disease, credits: 12, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:69,\n",
       " 'BS32707': Module: [code: BS32707, name: LSM3225: Molecular Microbiology in Human Diseases, credits: 12, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:70,\n",
       " 'BS32708': Module: [code: BS32708, name: LSM3243: Molecular Biophysics, credits: 12, scqflevel: 9, shelevel: 3, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:71,\n",
       " 'BS41004': Module: [code: BS41004, name: Research Project: Biological Sciences, credits: 40, scqflevel: 10, shelevel: 4, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:72,\n",
       " 'BS41005': Module: [code: BS41005, name: Research Skills in Biological Sciences, credits: 20, scqflevel: 10, shelevel: 4, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:73,\n",
       " 'BS41006': Module: [code: BS41006, name: Research Skills in Biomedical Sciences, credits: 20, scqflevel: 10, shelevel: 4, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:74,\n",
       " 'BS41007': Module: [code: BS41007, name: Research Project: Biomedical Sciences, credits: 40, scqflevel: 10, shelevel: 4, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:75,\n",
       " 'BS41009': Module: [code: BS41009, name: Internship/Work Placement, credits: 60, scqflevel: 10, shelevel: 4, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:76,\n",
       " 'BS41701': Module: [code: BS41701, name: LSM4199: Honours Project in Life Sciences, credits: 48, scqflevel: 10, shelevel: 4, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:77,\n",
       " 'BS41702': Module: [code: BS41702, name: LSM4213: System Neurobiology, credits: 12, scqflevel: 10, shelevel: 4, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:78,\n",
       " 'BS41703': Module: [code: BS41703, name: LSM4241: Functional Genomics, credits: 12, scqflevel: 10, shelevel: 4, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:79,\n",
       " 'BS41704': Module: [code: BS41704, name: LSM4245: Epigenetics and Chromatin Biology, credits: 12, scqflevel: 10, shelevel: 4, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:80,\n",
       " 'BS41705': Module: [code: BS41705, name: LSM4212: Pharmacogenetics and Drug Responses, credits: 12, scqflevel: 10, shelevel: 4, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:81,\n",
       " 'BS41706': Module: [code: BS41706, name: LSM4221: Drug Discovery and Clinical Trials, credits: 12, scqflevel: 10, shelevel: 4, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:82,\n",
       " 'BS42003': Module: [code: BS42003, name: Advanced Bioinformatics, credits: 15, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:83,\n",
       " 'BS42004': Module: [code: BS42004, name: Advanced Modern Drug Discovery, credits: 15, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:84,\n",
       " 'BS42005': Module: [code: BS42005, name: Advanced Plant Sciences, credits: 15, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:85,\n",
       " 'BS42006': Module: [code: BS42006, name: Advanced Immunology, credits: 15, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:86,\n",
       " 'BS42007': Module: [code: BS42007, name: Cancer Biology, credits: 15, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:87,\n",
       " 'BS42008': Module: [code: BS42008, name: Stem Cells in Development and Disease, credits: 15, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:88,\n",
       " 'BS42009': Module: [code: BS42009, name: Advanced Molecular Microbiology, credits: 15, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:89,\n",
       " 'BS42010': Module: [code: BS42010, name: Advanced Gene Regulation and Expression, credits: 15, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:90,\n",
       " 'BS42011': Module: [code: BS42011, name: Chemical Biology, credits: 15, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:91,\n",
       " 'BS42012': Module: [code: BS42012, name: Parasitology, credits: 15, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:92,\n",
       " 'BS42013': Module: [code: BS42013, name: Advanced Cell Signalling, credits: 15, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:93,\n",
       " 'BS42014': Module: [code: BS42014, name: Nutrients & Metabolic Disease, credits: 15, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:94,\n",
       " 'BS42017': Module: [code: BS42017, name: Translational Psychiatry & Neurology, credits: 15, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:95,\n",
       " 'BS42018': Module: [code: BS42018, name: Mapping Memories in the Brain, credits: 15, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:96,\n",
       " 'BS42019': Module: [code: BS42019, name: Cardiovascular Pharmacology, credits: 15, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:97,\n",
       " 'BS42021': Module: [code: BS42021, name: Heart & Circulation, credits: 15, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:98,\n",
       " 'BS42023': Module: [code: BS42023, name: Applied neuroanatomy and neurodegenerative disorders, credits: 15, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:99,\n",
       " 'BS42025': Module: [code: BS42025, name: The Conscious Brain: The Physiology & Pharmacology of Pain & Anaesthesia, credits: 15, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:100,\n",
       " 'BS42027': Module: [code: BS42027, name: Cancer Pharmacology and Treatment, credits: 15, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:101,\n",
       " 'BS42028': Module: [code: BS42028, name: Pharmacology of Treatment of Metabolic Disease, credits: 15, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:102,\n",
       " 'BS42701': Module: [code: BS42701, name: LSM4215: Extreme Physiology, credits: 12, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:103,\n",
       " 'BS42702': Module: [code: BS42702, name: LSM4217: Functional Ageing, credits: 12, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:104,\n",
       " 'BS42703': Module: [code: BS42703, name: LSM4225: Genetic Medicine in the Post-Genomic Era, credits: 12, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:105,\n",
       " 'BS42704': Module: [code: BS42704, name: LSM4232: Advanced Cell Biology, credits: 12, scqflevel: 10, shelevel: 4, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:106,\n",
       " 'BS51007': Module: [code: BS51007, name: Cell Biology and Disease, credits: 20, scqflevel: 11, shelevel: 5, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:107,\n",
       " 'BS51008': Module: [code: BS51008, name: Molecular Basis of Diseases: Genetics, Epigenetics and Genomics, credits: 20, scqflevel: 11, shelevel: 5, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:108,\n",
       " 'BS51009': Module: [code: BS51009, name: Scientific Method and Skills, credits: 20, scqflevel: 11, shelevel: 5, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:109,\n",
       " 'BS51010': Module: [code: BS51010, name: Scientific and Health Landscapes, credits: 20, scqflevel: 11, shelevel: 5, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:110,\n",
       " 'BS51011': Module: [code: BS51011, name: Writing and Communication Skills for Health and Science, credits: 20, scqflevel: 11, shelevel: 5, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:111,\n",
       " 'BS51012': Module: [code: BS51012, name: Tools, Software and Databases, credits: 20, scqflevel: 11, shelevel: 5, semester: 1] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:112,\n",
       " 'BS52007': Module: [code: BS52007, name: Advanced Technology Platforms, credits: 20, scqflevel: 11, shelevel: 5, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:113,\n",
       " 'BS52008': Module: [code: BS52008, name: MSc Project Proposal and Associated Skills, credits: 20, scqflevel: 11, shelevel: 5, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:114,\n",
       " 'BS52009': Module: [code: BS52009, name: Therapeutics, credits: 20, scqflevel: 11, shelevel: 5, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:115,\n",
       " 'BS52010': Module: [code: BS52010, name: Communication and Marketing Strategy, credits: 20, scqflevel: 11, shelevel: 5, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:116,\n",
       " 'BS52011': Module: [code: BS52011, name: Science and health communication industries and contemporary issues, credits: 20, scqflevel: 11, shelevel: 5, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:117,\n",
       " 'BS52012': Module: [code: BS52012, name: Digital and Visual Communication, credits: 20, scqflevel: 11, shelevel: 5, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:118,\n",
       " 'BS52013': Module: [code: BS52013, name: Biomolecular Modelling, credits: 20, scqflevel: 11, shelevel: 5, semester: 2] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:119,\n",
       " 'BS53003': Module: [code: BS53003, name: MSc Biobusiness Research Project, credits: 60, scqflevel: 11, shelevel: 5, semester: 3] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:120,\n",
       " 'BS53004': Module: [code: BS53004, name: MSc Research Project, credits: 60, scqflevel: 11, shelevel: 5, semester: 3] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:121,\n",
       " 'BS53005': Module: [code: BS53005, name: Science and Health Communication Masters Project, credits: 60, scqflevel: 11, shelevel: 5, semester: 3] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:122,\n",
       " 'BS53006': Module: [code: BS53006, name: Masters Placement: Science and Health Communication, credits: 60, scqflevel: 11, shelevel: 5, semester: 3] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:123,\n",
       " 'BS53007': Module: [code: BS53007, name: MSc Quantitative Biology Research Project, credits: 60, scqflevel: 11, shelevel: 5, semester: 3] ID: 4:04f9907a-1dd4-4d8a-a4fe-7ccb993179fa:124}"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "modules\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "f838fdbf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Applicability': 'From 2023-24',\n",
       " 'Section': 'Section 7: University management information (to be completed by the University’s Quality and Academic Standards office)',\n",
       " 'Heading': 'Details',\n",
       " '1.1': 'Bachelor of Science (Hons) in Biomedical Sciences',\n",
       " '1.2': 'Biomedical Sciences\\n(B1) Anatomy, physiology & pathology\\n(B2) Pharmacology, toxicology & pharmacy\\n(C0) Broadly-based programmes within biological sciences\\n(C1) Biology\\n(C9) Others in Biological Sciences\\n',\n",
       " '1.3': 'B900 Biomedical Sciences\\nB901 Biomedical Sciences with year in industry\\n',\n",
       " '1.4': 'University of Dundee',\n",
       " '1.5': 'University of Dundee',\n",
       " '1.6': 'English',\n",
       " '1.7': 'SCQF Level 10',\n",
       " '1.8': '',\n",
       " '1.9': '\\n\\n',\n",
       " '1.10': '\\n',\n",
       " '1.11': 'Details of the programme aims, indicative content and intended learning outcomes are provided in the Programme Specification.',\n",
       " '1.12': 'Details about the modules undertaken by the students are described in this Programme Specification, section 2.10',\n",
       " '1.13': 'The University’s grading scheme is described in the Assessment Policy for Taught Provision, available at http://www.dundee.ac.uk/pgla/policies/policy-taught-provision/.',\n",
       " '1.14': 'This qualification may lead to study at levels 11 and 12 of the Scottish Credit and Qualifications Framework.',\n",
       " '1.15': 'The programme is approved by the Royal Society of Biology accreditation scheme.',\n",
       " '2.1': 'With its strong emphasis on skills training and research led theory content, this programme aims to produce confident, self-reflective, critical-thinking Life Sciences graduates who are able to solve problems, and enthusiastically tackle new challenges that constantly arise with particular emphasis on the global context.',\n",
       " '2.2': 'Biomedical Sciences students focus on the molecular aspects of bioscience. Students study a common core curriculum at Levels 1 and 2 that provides a foundation of knowledge including numeracy/statistics, chemistry and molecular, physiological and pharmacological aspects of bioscience. In Level 3, Biomedical Science students specialise in neuroscience and/or physiological and/or pharmacological aspects of biomedicine covering the regulation of human organ systems, molecular and cellular physiology, receptor signalling processes and quantitative aspects of pharmacology.  Each theme is presented from a research-led perspective incorporating laboratory and analytical techniques that are used to research these topics. In addition to theory-based content, students undertake a semester-long practical module which delivers analytical experience in biomedical techniques that are common to the fields of neuroscience, pharmacology, physiology and which bears relevance to medical science. Students have a choice of modules relevant to their particular interests that are research-led and focus on latest developments in various areas of the biosciences actively researched in the School. In Level 4 students undertake a semester-long capstone project (either laboratory-based, data analysis with literature review, bioinformatics, science communication or bio-business). They also choose further research-led modules relevant to their interests and degree route. The choice of modules in Levels 3 and 4 together with the capstone project, are key in determining the degree specialism (see also Appendix for degree specialisms and their schedule of modules).  Students may take an optional Year in Industry between Levels 3 and 4.',\n",
       " '2.3': '1. Knowledge and understanding \\ndemonstrate an accurate understanding of the explanation of biomedical sciences phenomena at a variety of levels (from molecule to cell to organ to system function), with particular reference to the human body in health and disease\\nunderstand the common causes and effects of disease, the body’s defence mechanisms and approaches to treatment\\nexplain the relationship of evolutionary theory to their area of study\\nexplain the chemistry and structure of the major biological macromolecules and give detailed examples of how structure enables function\\ndemonstrate an understanding of the molecular basis of genetics and explain some detailed examples \\ndemonstrate critical knowledge and understanding of gene expression and how it is controlled with a detailed knowledge of specific examples and relevant experimental methods \\ndemonstrate knowledge of a wide range of cells (both prokaryotic and eukaryotic) and explain critically how their properties suit them for their biological function, and how they could be investigated experimentally \\ndevise and evaluate suitable experimental methods for the investigation of relevant areas of biomedical sciences\\ndemonstrate a critical understanding of molecular features and holistic roles of cells, tissues and organ systems in human health and disease (including knowledge and experience of some experimental techniques) \\n2. Subject-specific practical and intellectual skills and attributes\\ndemonstrate understanding of and competence in a broad range of practical techniques and skills relevant to the biosciences\\nplan, execute and present a piece of hypothesis-driven, independent work, demonstrating suitable time management and problem solving abilities, as well as interpretation and critical awareness of the quality of evidence \\nshow an appreciation of the ethical issues underpinning professional integrity and standards\\nbe able to construct reasoned arguments to support their position on the ethical and social impact of advances in the biosciences \\nrecord biological or biomedical data accurately and apply relevant advanced numerical skills to data \\ncommunicate science to peers and non-scientists \\ndemonstrate well developed strategies for updating, maintaining and enhancing their knowledge of the biosciences, including cross-disciplinary awareness \\naccess bioscience databases (e.g. small molecule or macromolecular databases) and use appropriate selection criteria to mine, manipulate, evaluate and interpret data.\\nAwareness and understanding of relevant health and safety standards and the hazards and risks associated with working in the laboratory, and an ability to apply safe working practise.\\n3. Transferable, employability and enterprise skills and attributes\\n•\\tWritten and oral communication of information and ideas\\n•\\tProblem solving, including the ability to think and conceptualise analytically, critically, creatively and laterally\\n•\\tTeamwork: the ability to establish working relationships with colleagues and work effectively as part of a team\\n•\\tTime management: the ability to manage your own time and set priorities\\n•\\tInformation Technology: using modern computing resources for data handling, information retrieval, document preparation, presentations and communication\\nAwareness and understanding of intellectual property issues and how they relate to the innovation process',\n",
       " '2.4': 'All modules are assessed by a combination of in-course and end-of-course procedures. Regular in-course assessments (e.g. practical reports, computer-based exercises, essays and data processing exercises) provide feedback on student progress and help students prepare for end-of-module examinations.\\nOn-line assignments are used extensively at Levels 1 and 2, with access via a PC on or off campus. Peer assessment operates in some team projects. The assessments aim to test both the knowledge gained, and the skills (including laboratory skills) which form an important part of the learning outcomes.\\nThe University of Dundee uses a criterion-based system on a 23 point scale. Assessment profiles for each module are inspected to see if they fit the general expectation, and may be moderated if (for instance) particular difficulties are identified post hoc with the match of questions and curricular content. Such moderation is overseen by the Programme Leads and Associate Dean Quality and Academic Standards, and approved by the External Examiner.',\n",
       " '2.5': 'The indicative assessment weightings are described below.  This may vary depending on module choices.\\n\\n',\n",
       " '2.6': 'The learning and teaching strategy underpinning the curriculum is to produce Biological and Biomedical Sciences undergraduate degree programmes which are:\\nFocused on laboratory skills, key principles, problem-solving and core knowledge.\\nDesigned to encourage independent ‘deep’ learning and critical thinking.\\nAligned with research, thereby making best use of the resource available in the School of Life Sciences to deliver world class undergraduate degrees.\\nUsing a common core curriculum at Levels 1 and 2, with the choice of more specialised modules at Levels 3 and 4 define the degree.\\nFully integrated across the programme to ensure that skills and understanding are appropriately reinforced without introducing redundancy.\\nMore competitive with respect to entry requirements for the vast majority of degrees, achievable by broadening the recruitment base and having more flexibility for Level 2 entry.\\nThis programme involves a wide\\xa0variety of teaching methods and approaches including lectures, tutorials, workshops, laboratory practicals, “flipped” classes and a capstone research project. Taught sessions\\xa0are supplemented\\xa0by individual reading and study.',\n",
       " '2.7': 'The indicative scheduled contact hours and guided independent study may vary depending on module choices.\\n\\n',\n",
       " '2.8': 'Teaching is carried out by staff from the Schools of Life Sciences and Medicine.  Full list of staff is available at http://www.lifesci.dundee.ac.uk/staff and http://medicine.dundee.ac.uk/staff',\n",
       " '2.9': 'The majority of teaching will take place at the city and Ninewells campuses at the University of Dundee; lecture rooms and rooms suitable for seminars/tutorials for small group teaching are required. Some teaching may be delivered at the Division of Plant Sciences, School of Life Sciences located at the James Hutton Institute.  A number of modules require the use of IT suites. Dedicated laboratories are required for practical modules. Projects will be accommodated within research laboratories across the Schools of Life Sciences and Medicine.  The optional Year in Industry Placement module will be located at partner industries or research organisations.',\n",
       " '2.10': '2.10 Structure of the programme\\nSchedule of modules\\nCore modules\\n* Students must choose a minimum of ONE and no more than TWO of these modules at level 4, subject to satisfactory pre-requisites.  \\n** Only for students taking the optional Year in Industry\\nOptional modules\\n',\n",
       " '2.11': 'Degree options: \\nBiomedical Sciences\\nBiomedical Sciences with year in industry\\n',\n",
       " '2.12': \"The University is committed to attracting and retaining students from under-represented groups and this module will make adjustments where it is reasonably possible to allow equality of opportunity for all students irrespective of age, disability, gender, race, religion or belief or sexual orientation.\\nContact for further information: School Disability Support Officer - Mrs Monica Lacey\\nor Life Sciences Athena SWAN representative - Dr Melissa D'Ascenzio\",\n",
       " '3.1': 'School of Life Sciences',\n",
       " '3.2': 'D’Arcy Thompson Unit',\n",
       " '3.3': 'School of Medicine',\n",
       " '3.4': 'Dr Stephen Kelley, s.p.kelley@dundee.ac.uk, +44 (0)1382 384760',\n",
       " '3.5': 'Dr Kim Dale, j.k.dale@dundee.ac.uk, +44 (0)1382 386290',\n",
       " '3.6': 'Ms Arlene Stewart, a.v.stewart@dundee.ac.uk, Head of Administration, +44 (0)1382 384970',\n",
       " '3.7': 'Dr Stephen Kelley, s.p.kelley@dundee.ac.uk, +44 (0)1382 384760',\n",
       " '4.1': 'SCQF, QAA Benchmarks http://www.qaa.ac.uk/assuring-standards-and-quality\\nFor Biosciences:  https://www.qaa.ac.uk/docs/qaa/subject-benchmark-statements/subject-benchmark-statement-biosciences.pdf\\nFor Biomedical Sciences:  https://www.qaa.ac.uk/docs/qaa/subject-benchmark-statements/subject-benchmark-statement-biomedical-sciences.pdf\\nThe Royal Society of Biology Degree Accreditation https://www.rsb.org.uk/education/accreditation',\n",
       " '4.2': 'All members of staff have either a PhD in a relevant area and/or professional qualifications. Teaching staff have a range of teaching and research experience, and are able to bring expertise covering a variety of modern biological and biomedical topics to their teaching. Several are highly experienced teachers, and have close familiarity with a range of assessment and delivery styles. Less experienced members of the teaching team are provided with guidance on all aspects of module teaching by the more experienced members of the team. \\nTeaching staff will be supported in assurance of academic standards via oversight of all quality issues by the Quality and Academic Standards Committees of the School of Life Sciences. Staff will take advantage of professional development opportunities offered through the Academic Skills Centre, and the School of Life Sciences internal teaching seminar programme.',\n",
       " '4.3': 'N/A',\n",
       " '4.4': 'The programme is accredited by the Royal Society of Biology.',\n",
       " '4.5': 'N/A',\n",
       " '4.6': 'Exchanges: Students are eligible for international exchange\\nPlacements: Only for those students taking the optional Year in Industry. Students will undertake the Life Sciences Placement module. Students may be on placement with a range of employers at diverse locations. The suitability of an employer to act as a placement provider will be established and approved in line with the University’s Code of Practice on Student Placements and Exchanges, coordinated by a dedicated member of academic staff (Module Leader for Life Sciences Placement module). An appropriate risk assessment of placement activities will be undertaken and recorded prior to each placement. The responsibilities of both the University and the placement provider will be defined in writing. Clearly defined points of contact and lines of communication between School and placement provider will be put in place. The School will maintain a record of all placements. Students will be briefed prior to commencing their placement activities. The student’s participation and welfare will be monitored regularly via direct communication with the student – a record of which will be kept for future scrutiny. Monitoring and review of placements will be undertaken on a yearly basis.',\n",
       " '4.7': \"All staff are committed to providing an inclusive curriculum including access to a full teaching and learning programme for disabled students. The School liaises with the University's Equality and Diversity Officer and the Head of Disability Services to determine the most appropriate way to achieve this and ensure that no group or individual is disadvantaged by the learning environment. School and Programme Disability Support Officers support this process and ensure that specific adjustments are arranged to meet individual disabled students' needs. Further details of the University's provision for disabled students are available at:  http://www.dundee.ac.uk/studentservices/disabilityservices/\",\n",
       " '4.8': '1. Staff teaching on the module use the Reading List facility, to ensure that there are adequate resources in the Library for this programme.\\n2. My Dundee (VLE) is used extensively to support teaching and learning.\\n3. Exams, with a single exception, are online, making use of QuestionMark Perception, ExamOnline and other suitable software/systems as appropriate. The exception is paper based due to the requirement to draw complex chemical structures and reactions.\\n4. Microsoft Office, statistical software (R Studio) and other specialist software (Flowing, Jupyter Notebooks, UCSF Chimera, Jalview, ImageJ) is required for some of the modules.',\n",
       " '4.9': 'Produced annually through the allocation of workload and resource. The programme will be delivered through the Schools of Life Sciences and Medicine.',\n",
       " '5.1': 'Current',\n",
       " '5.2': 'N/A',\n",
       " '5.3': 'Level 1, 2 and 3',\n",
       " '5.4': 'BSc Honours Biomedical Sciences\\nBSc Honours Biomedical Sciences (with a year in industry)\\n\\nBSc Ordinary Biomedical Sciences\\nDiploma of Higher Education\\nCertificate of Higher Education',\n",
       " '5.5': 'N/A',\n",
       " '5.6': 'One, in September',\n",
       " '5.7': 'Entry with advanced standing will require that students have successfully fulfilled courses equivalent to those within the programme, to satisfy the University’s recognition of prior learning credits requirements.',\n",
       " '5.8': 'As set by the University',\n",
       " '5.9': 'As set by the University',\n",
       " '5.10': 'As set by the University',\n",
       " '5.11': 'RUK scholarships\\t\\t\\t\\nRUK bursaries\\t\\t\\t\\nInternational scholarships\\t\\t\\nRUK scholarships:\\nhttps://www.dundee.ac.uk/media/dundeewebsite/study/scholarships/ruk/Terms-and-Conditions-for-Rest-of-UK-Scholarships-and-Bursaries-2017.pdf\\n\\nAcademic excellence scholarship: Entry qualifications of AAB or above at A-Level; £3,000 per year of study:\\nhttps://www.dundee.ac.uk/study/scholarships/2018-sept/academic-excellence-ruk/\\n\\nRest of UK bursary: Have a residual household income of less than £20,000 per annum.\\nhttps://www.dundee.ac.uk/study/scholarships/2018-sept/ruk-bursary/ \\n\\nDiscover Dundee: Eligibility: Be classified as RUK fee status by the University of Dundee: £2k 1st year only\\nhttps://www.dundee.ac.uk/study/scholarships/2018-sept/discover-dundee-ruk/:\\n\\nInternational scholarship:\\nThe University of Dundee Global Excellence Scholarship (Undergraduate): \\nTier1: Up to £25,000 (£5,000 per year of study): achieve the equivalent of UK A-Levels at AAB or above\\nTier 2: less than UK A-Levels AAB, we will review your application and may invite you to apply for the scholarship. The value of the scholarship for application is £5,000 in entry year only. In general, applications will be considered if your qualifications are UK A-Level BBB or above, however we will review all applications on an individual basis and a decision will be made by our scholarship panel.\\nhttps://www.dundee.ac.uk/study/scholarships/2018-sept/uod-global-excellence-ug/\\n\\t',\n",
       " '5.12': 'Host country government scholarships',\n",
       " '5.13': 'Apart from textbooks and graduation fees, additional costs are those to cover purchase of a suitable lab coat.',\n",
       " '5.14': 'UCAS\\t\\t\\t\\t\\nOther*\\t\\t\\t\\t',\n",
       " '5.15': 'Undergraduate',\n",
       " '5.16': 'Students must apply through UCAS providing the details required.',\n",
       " '5.17': '\\n\\n',\n",
       " '5.18': 'All programmes are assigned to a reporting group for the purpose of data analyses and reporting.  ASRS will allocate the programme to a relevant SITS group as part of the approval process based on information that is provided.\\nAdvice will be provided by ASRS.',\n",
       " '5.19': 'N/A',\n",
       " 'The': 'The template developed by ASRS is available at:\\nhttp://uod.ac.uk/update-my-course\\nIt must be completed and accompany the programme documentation as part of the new programme approval process.',\n",
       " 'Version': 'QASv6 – 21/01/23',\n",
       " 'Status': 'Revise schedule of modules to remove:\\nBS11001 Biomolecules, stage 1, (SCQF level 7); SCQF 10 credits\\nBS11002 Evolution of Cells, stage 1, (SCQF level 7); SCQF 10 credits \\nBS12001 Genes and Hereditary, stage 1, (SCQF level 7); SCQF 10 credits\\nBS12002 Origins of Animal Organ Systems, stage 1, (SCQF level 7); SCQF 10 credits',\n",
       " 'Date': 'QASC 09/02/23'}"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "{s:structure[s][-1].text for s in structure}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "2116a6aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "programmes={'BIMS':factory.get_or_create_Element('Programme', code='BIMS',name=structure['1.1'][-1].text)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "0c841729",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on Programme in module __main__ object:\n",
      "\n",
      "class Programme(Node)\n",
      " |  Programme(factory, **params)\n",
      " |  \n",
      " |  Represents a Node in a Neo4j database. \n",
      " |  Contains methods for accessing links, ensuring that on creation the \n",
      " |  correct parameters are given etc.\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      Programme\n",
      " |      Node\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, factory, **params)\n",
      " |      Create a Programme instance.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      factory - database connection and cache for Programme\n",
      " |      **params : \n",
      " |          Create or replace table Programme (\n",
      " |  \n",
      " |  loadILO(self)\n",
      " |      Loads ILO links from the database. Does not create new links or ILOs.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      None.\n",
      " |  \n",
      " |  loadmodules(self)\n",
      " |      Loads module links from the database to the modules list. Does not create now entries in the database.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      None.\n",
      " |  \n",
      " |  map_ilo(self, ilo, year, remove=False)\n",
      " |      Associates or updates a Programme ILO mapping to a Programme.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      ilo : ProgrammeILO \n",
      " |          Programme ILO object\n",
      " |      year : text describing the academic year\n",
      " |          Must be either the start year, or for removing an ILO the last year for which it will be relevant.\n",
      " |      remove : Boolean, optional\n",
      " |          Flag to say whether the ILO should be dissociated from teh Programme. The default is False.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      None.\n",
      " |  \n",
      " |  map_module(self, module, optional=0, year=None, remove=False)\n",
      " |      Add a module to the Programme, updating if exisitng  already appended.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      module : Module object\n",
      " |          DESCRIPTION.\n",
      " |      optional: Boolean\n",
      " |      remove: Academic Year for last instance\n",
      " |      Returns\n",
      " |      -------\n",
      " |      None.\n",
      " |  \n",
      " |  withdraw(self, year)\n",
      " |      Set the Programme to WITHDRAWN. Only possible if there are no future versions.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      None.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  ARCHIVED = 2\n",
      " |  \n",
      " |  CURRENT = 1\n",
      " |  \n",
      " |  DRAFT = 0\n",
      " |  \n",
      " |  WITHDRAWN = 3\n",
      " |  \n",
      " |  ntype = 'Programme'\n",
      " |  \n",
      " |  optionalParams = {'endyear': 'Last academic year', 'school': 'School w...\n",
      " |  \n",
      " |  requiredParams = {'code': 'Programme code', 'name': 'Programme name'}\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from Node:\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return repr(self).\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Returns a string represntation of the node\n",
      " |  \n",
      " |  getEdges(self, relation=None, **kwargs)\n",
      " |      retrieves edges connected to or from the Node.\n",
      " |      Relation is the type of relation to retrieve (default is all)\n",
      " |      other arguments are applied as matching parameters\n",
      " |  \n",
      " |  getparam(self, prop, default=None)\n",
      " |      Returns a parameter value (default if not found)\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      prop : property name\n",
      " |          \n",
      " |      default : TYPE, optional\n",
      " |           The default is None.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Property value\n",
      " |  \n",
      " |  isCurrent(self, year)\n",
      " |      If the edge has a start/end year, check if the given year is the period in which the relation is valid\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Boolean: True if the eyar is between start and end (inclusive), after start if no end, or True if no start\n",
      " |  \n",
      " |  setparam(self, key, value)\n",
      " |      Sets the parameter key to value. No type checking.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      key : legal value for a dictionary key\n",
      " |          \n",
      " |      value : Value to store.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      None.\n",
      " |  \n",
      " |  toDict(self)\n",
      " |  \n",
      " |  update(self, **kwargs)\n",
      " |      Update named parameters for a Node\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      **kwargs : key,value parameter set\n",
      " |          any key,value parameters. Existing values will be overwritten.\n",
      " |          No type checking\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      None.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from Node:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(programmes['BIMS'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "1ca8141c",
   "metadata": {},
   "outputs": [],
   "source": [
    "for m in modmap:\n",
    "    if modules.get(m[1].strip('*')):\n",
    "        programmes['BIMS'].map_module(modules[m[1].strip('*')],m[0]=='Elective', year='23/24')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "50a74bad",
   "metadata": {},
   "outputs": [],
   "source": [
    "docs = {\n",
    "    'BIO':'BSc Hons Biologsci Programme Specification QASv3.docx',\n",
    "'BIMS': 'BSc Hons Biomedical Sciences Programme Specification 2324.docx',\n",
    " 'NEUR': 'BSc Hons Neuroscience Programme Specification QASv8.docx',\n",
    " 'PHAR': 'BSc Hons Pharmacology Programme Specification QASv6.docx',\n",
    " 'PHYS': 'BSc Hons Physiological Sciences Programme Specification QASv6.docx'\n",
    "    \n",
    "}\n",
    "    \n",
    "\n",
    "progdir='c:/Users/marti/Documents/LifeSciteaching/Curriculum/Programmes/UG/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "06c2b05e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Core', 'BS11009', 'The Building Blocks of Life', '1', '7', '20', 'SLS']\n",
      "['Core', 'BS11008', 'Core Skills in the Life Sciences 1A', '1', '7', '20', 'SLS']\n",
      "['Core', 'BS12011', 'Building the Organism', '1', '7', '20', 'SLS']\n",
      "['Core', 'BS12010', 'Core Skills in the Life Sciences 1B', '1', '7', '20', 'SLS']\n",
      "['Core', 'BS21001', 'Statistics and Experimental Design', '2', '8', '10', 'SLS']\n",
      "['Core', 'BS21002', 'Cellular Communication', '2', '8', '10', 'SLS']\n",
      "['Core', 'BS21012', 'Core Skills in the Life Sciences 2A', '2', '8', '20', 'SLS']\n",
      "['Core', 'BS22001', 'Human Physiology and Pharmacology', '2', '8', '20', 'SLS']\n",
      "['Core', 'BS22002', 'Biomolecular Mechanisms', '2', '8', '20', 'SLS']\n",
      "['Core', 'BS22003', ' Core Skills in the Life Sciences 2B', '2', '8', '20', 'SLS']\n",
      "['Core', 'BS31013', 'Biomembranes', '3', '9', '15', 'SLS/SoM']\n",
      "['Core', 'BS31016', 'Practical Techniques in Biomedical Sciences', '3', '9', '15', 'SLS/SoM']\n",
      "['Core', 'BS31019', 'Regulatory Physiology & Pharmacology', '3', '9', '15', 'SLS/SoM']\n",
      "['Core', 'BS30007**', 'Life Sciences Placement', '3', '9', '80', 'SLS ']\n",
      "['Core', 'BS41006', 'Research Skills – Biomedical Sciences: Biological Sciences', '4', '10', '20', 'SLS/SoM']\n",
      "['Core', 'BS41007', 'Research project – Biomedical Sciences', '4', '10', '40', 'SLS/SoM']\n",
      "['Elective', 'BS11005', 'Numeracy, Chemistry and Physics for the Biological and Biomedical Sciences', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS11006', 'The Poison Pen', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS12005', 'Science and Society', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS12008', 'Introduction to Scientific Enterprise', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS12009', 'Fundamentals of Organic Chemistry', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS21008', 'Organic Chemistry for the Life Sciences', '2', '8', '20', 'SLS']\n",
      "['Elective', 'BS21009', 'Data Analysis for the Life Sciences', '2', '8', '20', 'SLS']\n",
      "['Elective', 'BS21010', 'Introductory Programming for the Life Sciences', '2', '8', '20', 'SLS']\n",
      "['Elective', 'BS21011', 'Developing Scientific Enterprise', '2', '8', '20', 'SLS']\n",
      "['Elective', 'BS31004', 'Biochemistry and Cell Biology', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS31006', 'Gene Regulation and Expression', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS31020', 'Experimental Cell Culture ', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32006', 'Cell Signalling*', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32026', 'Sensory & Motor Neuroscience', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32028', 'Molecular Pharmacology', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS32029', 'Cell Proliferation and Survival Mechanisms Underlying Disease', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS32030', 'Data and Statistical Analysis*', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32031', 'Bioinformatics Research Skills I*', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32033', 'Epidemiology and Health Informatics', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS32005', 'Cell and Developmental Biology*', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32009', 'Immunology*', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32020', 'Human Epithelial Biology', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32021', 'Quantitative Pharmacology', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32024', 'Behavioural Neuroscience and Pharmacology', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS32032', 'Science Communication', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42003', 'Advanced Bioinformatics', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS42006', 'Advanced Immunology*', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS42007', 'Cancer Biology', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42008', 'Stem Cells in Development and Disease*', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS42009', 'Advanced Molecular Microbiology*', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS42012', 'Parasitology*', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS42013', 'Advanced Cell Signalling*', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS42014', 'Nutrients & Metabolic Disease', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42017', 'Translational Psychiatry & Neurology', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42018', 'Synaptic Plasticity & Cognition', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42019', 'Cardiovascular Pharmacology', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42021', 'Heart & Circulation', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42023', 'Applied Neuroanatomy & Neurodegenerative Disorders', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42025', 'The Conscious Brain: The Physiology & Pharmacology of Pain and Anaesthesia', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42027', 'Cancer Pharmacology & Treatment', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42028', 'Pharmacology and Treatment of Metabolic Disease ', '4', '10', '15', 'SLS/SoM']\n",
      "['Core', 'BS11009', 'The Building Blocks of Life', '1', '7', '20', 'SLS']\n",
      "['Core', 'BS11008', 'Core Skills in the Life Sciences 1A', '1', '7', '20', 'SLS']\n",
      "['Core', 'BS12011', 'Building the Organism', '1', '7', '20', 'SLS']\n",
      "['Core', 'BS12010', 'Core Skills in the Life Sciences 1B', '1', '7', '20', 'SLS']\n",
      "['Core', 'BS21001', 'Statistics and Experimental Design', '2', '8', '10', 'SLS']\n",
      "['Core', 'BS21002', 'Cellular Communication', '2', '8', '10', 'SLS']\n",
      "['Core', 'BS21012', 'Core Skills in the Life Sciences 2A', '2', '8', '20', 'SLS']\n",
      "['Core', 'BS22001', 'Human Physiology and Pharmacology', '2', '8', '20', 'SLS']\n",
      "['Core', 'BS22002', 'Biomolecular Mechanisms', '2', '8', '20', 'SLS']\n",
      "['Core', 'BS22003', 'Core Skills in the Life Sciences 2B', '2', '8', '20', 'SLS']\n",
      "['Core', 'BS31013', 'Biomembranes', '3', '9', '15', 'SLS/SoM']\n",
      "['Core', 'BS31016', 'Practical Techniques in Biomedical Sciences', '3', '9', '15', 'SLS/SoM']\n",
      "['Core', 'BS31019', 'Regulatory Physiology & Pharmacology', '3', '9', '15', 'SLS/SoM']\n",
      "['Core', 'BS32024', '\\nBehavioural Neuroscience and Pharmacology', '3', '9', '15', 'SLS/SoM']\n",
      "['Core', 'BS32026', 'Sensory & Motor Neuroscience', '3', '9', '15', 'SLS/SoM']\n",
      "['Core', 'BS30007**', 'Life Sciences Placement', '3', '9', '80', 'SLS ']\n",
      "['Core', 'BS41006', 'Research Skills – Biomedical Sciences: Biological Sciences', '4', '10', '20', 'SLS/SoM']\n",
      "['Core', 'BS41007', 'Research project – Biomedical Sciences', '4', '10', '40', 'SLS/SoM']\n",
      "['Core', 'BS42017', 'Translational Psychiatry & Neurology', '4', '10', '15', 'SLS/SoM']\n",
      "['Core', 'BS42023', 'Applied Neuroanatomy & Neurodegenerative Disorders', '4', '10', '15', 'SLS/SoM']\n",
      "['Core', 'BS42025', 'The Conscious Brain: The Physiology & Pharmacology of Pain and Anaesthesia', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS11005', 'Numeracy, Chemistry and Physics for the Biological and Biomedical Sciences', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS11006', 'The Poison Pen', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS12005', 'Science and Society', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS12008', 'Introduction to Scientific Enterprise', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS12009', 'Fundamentals of Organic Chemistry', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS21008', 'Organic Chemistry for the Life Sciences', '2', '8', '20', 'SLS']\n",
      "['Elective', 'BS21009', 'Data Analysis for the Life Sciences', '2', '8', '20', 'SLS']\n",
      "['Elective', 'BS21010', 'Introductory Programming for the Life Sciences', '2', '8', '20', 'SLS']\n",
      "['Elective', 'BS21011', 'Developing Scientific Enterprise', '2', '8', '20', 'SLS']\n",
      "['Elective', 'BS31004', 'Biochemistry and Cell Biology', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS31006', 'Gene Regulation and Expression', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS31020', 'Experimental Cell Culture ', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32004', 'Molecular Microbiology', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32006', 'Cell Signalling', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32028', 'Molecular Pharmacology', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS32029', 'Cell Proliferation and Survival Mechanisms Underlying Disease', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS32030', 'Data and Statistical Analysis', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32031', 'Bioinformatics Research Skills I', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32033', 'Epidemiology and Health Informatics', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS32005', 'Cell and Developmental Biology', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32009', 'Immunology', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32020', 'Human Epithelial Biology', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32021', 'Quantitative Pharmacology', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32032', 'Science Communication', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42018', 'Mapping Memories in the Brain', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42003', 'Advanced Bioinformatics', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS42006', 'Advanced Immunology', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS42007', 'Cancer Biology', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42008', 'Stem Cells in Development and Disease', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS42009', 'Advanced Molecular Microbiology', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS42012', 'Parasitology', '4', '10', '15', 'SLS']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Elective', 'BS42013', 'Advanced Cell Signalling', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS42014', 'Nutrients & Metabolic Disease', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42019', 'Cardiovascular Pharmacology', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42021', 'Heart & Circulation', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42027', 'Cancer Pharmacology & Treatment', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42028', 'Pharmacology and Treatment of Metabolic Disease ', '4', '10', '15', 'SLS/SoM']\n",
      "['Core', 'BS11009', 'The Building Blocks of Life', '1', '7', '20', 'SLS']\n",
      "['Core', 'BS11008', 'Core Skills in the Life Sciences 1A', '1', '7', '20', 'SLS']\n",
      "['Core', 'BS12011', 'Building the Organism', '1', '7', '20', 'SLS']\n",
      "['Core', 'BS12010', 'Core Skills in the Life Sciences 1B', '1', '7', '20', 'SLS']\n",
      "['Core', 'BS21001', 'Statistics and Experimental Design', '2', '8', '10', 'SLS']\n",
      "['Core', 'BS21002', 'Cellular Communication', '2', '8', '10', 'SLS']\n",
      "['Core', 'BS21012', 'Core Skills in the Life Sciences 2A', '', '', '', '']\n",
      "['Core', 'BS22001', 'Human Physiology and Pharmacology', '2', '8', '20', 'SLS']\n",
      "['Core', 'BS22002', 'Biomolecular Mechanisms', '2', '8', '20', 'SLS']\n",
      "['Core', 'BS22003', ' Core Skills in the Life Sciences 2B', '2', '8', '20', 'SLS']\n",
      "['Core', 'BS31013', 'Biomembranes', '3', '9', '15', 'SLS/SoM']\n",
      "['Core', 'BS31016', 'Practical Techniques in Biomedical Sciences', '3', '9', '15', 'SLS/SoM']\n",
      "['Core', 'BS31019', 'Regulatory Physiology & Pharmacology', '3', '9', '15', 'SLS/SoM']\n",
      "['Core', 'BS32021', 'Quantitative Pharmacology', '3', '9', '15', 'SLS']\n",
      "['Core', 'BS32028', 'Molecular Pharmacology', '3', '9', '15', 'SLS/SoM']\n",
      "['Core', 'BS30007**', 'Life Sciences Placement', '3', '9', '80', 'SLS ']\n",
      "['Core', 'BS41006', 'Research Skills – Biomedical Sciences: Biological Sciences', '4', '10', '20', 'SLS/SoM']\n",
      "['Core', 'BS41007', 'Research project – Biomedical Sciences', '4', '10', '40', 'SLS/SoM']\n",
      "['Core', 'BS42014', 'Nutrients & Metabolic Disease*', '4', '10', '15', 'SLS/SoM']\n",
      "['Core', 'BS42019', 'Cardiovascular Pharmacology*', '4', '10', '15', 'SLS/SoM']\n",
      "['Core', 'BS42021', 'Heart & Circulation*', '4', '10', '15', 'SLS/SoM']\n",
      "['Core', 'BS42025', 'The Conscious Brain: The Physiology & Pharmacology of Pain & Anaesthesia*', '4', '10', '15', 'SLS/SoM']\n",
      "['Core', 'BS42027', 'Cancer Pharmacology & Treatment*', '4', '10', '15', 'SLS/SoM']\n",
      "['Core', 'BS42028', 'Pharmacology & Treatment of Metabolic Disease*', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS11005', 'Numeracy, Chemistry and Physics for the Biological and Biomedical Sciences', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS11006', 'The Poison Pen', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS12005', 'Science and Society', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS12008', 'Introduction to Scientific Enterprise', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS12009', 'Fundamentals of Organic Chemistry', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS21008', 'Bioinorganic and Organic Chemistry for the Life Sciences Workshops', '2', '8', '20', 'SLS']\n",
      "['Elective', 'BS21009', 'Data Analysis for the Life Sciences', '2', '8', '20', 'SLS']\n",
      "['Elective', 'BS21010', 'Introductory Programming for the Life Sciences', '2', '8', '20', 'SLS']\n",
      "['Elective', 'BS21011', 'Developing Scientific Enterprise', '2', '8', '20', 'SLS']\n",
      "['Elective', 'BS31004', 'Biochemistry and Cell Biology', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS31006', 'Gene Regulation and Expression', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS31020', 'Experimental Cell Culture ', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32005', 'Cell & Developmental Biology', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32006', 'Cell Signalling', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32009', 'Immunology', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32020 ', 'Human Epithelial Biology', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32024', 'Behavioural Neuroscience and Pharmacology', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS32026', 'Sensory and Motor Neuroscience', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32029', 'Cell Proliferation and Survival Mechanisms Underlying Disease', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS32030', 'Data and Statistical Analysis', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32031', 'Bioinformatics Research Skills I', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32032', 'Science Communication', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42003', 'Advanced Bioinformatics', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS32033', 'Epidemiology and Health Informatics', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42006', 'Advanced Immunology', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS42007', 'Cancer Biology', '4', '10', '15', 'SLS/SOM']\n",
      "['Elective', 'BS42008', 'Stem Cells in Development and Disease', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS42017', 'Translational Psychiatry and Neurology', '4', '10', '15', 'SLS/SoM']\n",
      "['Core', 'BS11009', 'The Building Blocks of Life', '1', '7', '20', 'SLS']\n",
      "['Core', 'BS11008', 'Core Skills in the Life Sciences 1A', '1', '7', '20', 'SLS']\n",
      "['Core', 'BS12011', 'Building the Organism', '1', '7', '20', 'SLS']\n",
      "['Core', 'BS12010', 'Core Skills in the Life Sciences 1B', '1', '7', '20', 'SLS']\n",
      "['Core', 'BS21001', 'Statistics and Experimental Design', '2', '8', '10', 'SLS']\n",
      "['Core', 'BS21002', 'Cellular Communication', '2', '8', '10', 'SLS']\n",
      "['Core', 'BS21012', 'Core Skills in the Life Sciences 2A', '2', '8', '20', 'SLS']\n",
      "['Core', 'BS22001', 'Biomedical Sciences', '2', '8', '20', 'SLS']\n",
      "['Core', 'BS22002', 'Biomolecular Mechanisms', '2', '8', '20', 'SLS']\n",
      "['Core', 'BS22003', 'Core Skills in the Life Sciences 2B', '2', '8', '20', 'SLS']\n",
      "['Core', 'BS31013', 'Biomembranes', '3', '9', '15', 'SLS/SoM']\n",
      "['Core', 'BS31016', 'Practical Techniques in Biomedical Sciences', '3', '9', '15', 'SLS/SoM']\n",
      "['Core', 'BS31019', 'Regulatory Physiology & Pharmacology', '3', '9', '15', 'SLS/SoM']\n",
      "['Core', 'BS32020', 'Human Epithelial Biology', '3', '9', '15', 'SLS']\n",
      "['Core', 'BS32026', 'Sensory & Motor Neuroscience', '3', '9', '15', 'SLS']\n",
      "['Core', 'BS30007**', 'Life Sciences Placement', '3', '9', '80', 'SLS ']\n",
      "['Core', 'BS41006', 'Research Skills – Biomedical Sciences: Biological Sciences', '4', '10', '20', 'SLS/SoM']\n",
      "['Core', 'BS41007', 'Research project – Biomedical Sciences', '4', '10', '40', 'SLS/SoM']\n",
      "['Core', 'BS42014', 'Nutrients & Metabolic Disease*', '4', '10', '15', 'SLS/SoM']\n",
      "['Core', 'BS42021', 'Heart & Circulation*', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS11005', 'Numeracy, Chemistry and Physics for the Biological and Biomedical Sciences', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS11005', 'The Poison Pen', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS12005', 'Science and Society', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS12008', 'Introduction to Scientific Enterprise', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS12009', 'Fundamentals of Organic Chemistry', '1', '7', '20', 'SLS']\n",
      "['Elective', 'BS21008', 'Organic Chemistry for the Life Sciences', '2', '8', '20', 'SLS']\n",
      "['Elective', 'BS21009', 'Data Analysis for the Life Sciences', '2', '8', '20', 'SLS']\n",
      "['Elective', 'BS21010', 'Introductory Programming for the Life Sciences', '2', '8', '20', 'SLS']\n",
      "['Elective', 'BS21011', 'Developing Scientific Enterprise', '2', '8', '20', 'SLS']\n",
      "['Elective', 'BS31004', 'Biochemistry and Cell Biology', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS31006', 'Gene Regulation and Expression', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS31020', 'Experimental Cell Culture ', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32004', 'Molecular Microbiology', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32005', 'Cell and Developmental Biology', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32006', 'Cell Signalling', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32009', 'Immunology', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32021', 'Quantitative Pharmacology', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32024', 'Behavioural Neuroscience and Pharmacology', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS32028', 'Molecular Pharmacology', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS32029', 'Cell Proliferation and Survival Mechanisms Underlying Disease', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS32030', 'Data and Statistical Analysis', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32031', 'Bioinformatics Research Skills I', '3', '9', '15', 'SLS']\n",
      "['Elective', 'BS32032', 'Science Communication', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS32033', 'Epidemiology and Health Informatics', '3', '9', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42006', 'Advanced Immunology*', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS42008', 'Stem Cells in Development and Disease*', '4', '10', '15', 'SLS']\n",
      "['Elective', 'BS42017', 'Translational Psychiatry & Neurology', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42019', 'Cardiovascular Pharmacology', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42023', 'Applied Neuroanatomy & Neurodegenerative Disorders', '4', '10', '15', 'SLS/SoM']\n",
      "['Elective', 'BS42028', 'Pharmacology and Treatment of Metabolic Disease ', '4', '10', '15', 'SLS/SoM']\n"
     ]
    }
   ],
   "source": [
    "for route in ['BIMS','NEUR','PHAR','PHYS']:\n",
    "    doc = docx.Document(os.path.join(progdir,docs[route]))\n",
    "    modmap = []\n",
    "    structure ={}\n",
    "    for t in doc.tables:\n",
    "        for r in range(len(t.rows)):\n",
    "            #print(len(t.rows), len(t.row_cells(r)), t.row_cells(r)[0].text,[c.tables for c in t.row_cells(r)])\n",
    "            structure[t.row_cells(r)[0].text.split()[0]] = t.row_cells(r)\n",
    "    tc=0\n",
    "    for m in structure['2.10'][0].tables:\n",
    "        coretype=['Core','Elective'][tc%2]\n",
    "        tc+=1\n",
    "        for r in range(1,len(m.rows)):\n",
    "            print([coretype]+[c.text for c in m.row_cells(r)])\n",
    "            modmap.append([coretype]+[c.text for c in m.row_cells(r)])\n",
    "    programmes[route]=factory.get_or_create_Element('Programme', code=route,name=structure['1.1'][-1].text)\n",
    "    for m in modmap:\n",
    "        if modules.get(m[1].strip('*')):\n",
    "            programmes[route].map_module(modules[m[1].strip('*')],m[0]=='Elective', year='23/24')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
