{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1140bf43",
   "metadata": {},
   "outputs": [],
   "source": [
    "import docx\n",
    "import os\n",
    "import re\n",
    "from CurriculumDB.Modelsn4j import *\n",
    "from neo4j import GraphDatabase\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "46d6f748",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<neo4j._sync.driver.BoltDriver object at 0x00000255A26BA350>\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "# URI examples: \"neo4j://localhost\", \"neo4j+s://xxx.databases.neo4j.io\"\n",
    "URI = \"bolt://localhost:7687\"\n",
    "AUTH = (\"curriculum\", \"mycurriculum\")\n",
    "\n",
    "driver = GraphDatabase.driver(URI, auth=AUTH)\n",
    "print(driver)\n",
    "print(driver.verify_connectivity())\n",
    "#help(driver.verify_connectivity)\n",
    "factory = CurriculumFactory(driver, 'curriculumdb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "926b73d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "help(factory.get_or_create_Element)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6b3a10b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "moduledir = \"C:/Users/dmamartin/OneDrive - University of Dundee/Teaching/CurriculumDB/Modules/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e9e92987",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Looking in Stage 1\n",
      "Checking ARCHIVE\n",
      "Checking BS10001 Widening Access Programme Collective Module Specification QASv1.docx\n",
      "processing BS10001 Widening Access Programme Collective Module Specification QASv1.docx\n",
      "Checking BS11005 Numeracy, Chemistry and Physics for the Biological and Biomedical Sciences Spec QASv4.docx\n",
      "processing BS11005 Numeracy, Chemistry and Physics for the Biological and Biomedical Sciences Spec QASv4.docx\n",
      "Checking BS11006 The Poison Pen Specification QASv4.docx\n",
      "processing BS11006 The Poison Pen Specification QASv4.docx\n",
      "Checking BS11008 Core Skills in the Life Sciences 1A QASv3.docx\n",
      "processing BS11008 Core Skills in the Life Sciences 1A QASv3.docx\n",
      "Checking BS11009 The Building Blocks of Life QASv2.docx\n",
      "processing BS11009 The Building Blocks of Life QASv2.docx\n",
      "Checking BS12005 Science and Society Module Spec QASv4.docx\n",
      "processing BS12005 Science and Society Module Spec QASv4.docx\n",
      "Checking BS12008 Introduction to Scientific Enterprise Specification QASv4.docx\n",
      "processing BS12008 Introduction to Scientific Enterprise Specification QASv4.docx\n",
      "Checking BS12009 Fundamentals of Organic Chemistry Module Specification QASv1.docx\n",
      "processing BS12009 Fundamentals of Organic Chemistry Module Specification QASv1.docx\n",
      "Checking BS12010 Core Skills in the Life Sciences 1B Specification QASv2.docx\n",
      "processing BS12010 Core Skills in the Life Sciences 1B Specification QASv2.docx\n",
      "Checking BS12011 Building the Organism QASv3.docx\n",
      "processing BS12011 Building the Organism QASv3.docx\n",
      "Checking BS12701 Singapore Studies Module Specification QASv1.docx\n",
      "processing BS12701 Singapore Studies Module Specification QASv1.docx\n",
      "Checking BS12702 General Education Module Specification QASv1.docx\n",
      "processing BS12702 General Education Module Specification QASv1.docx\n",
      "Checking Withdrawn\n",
      "Looking in Stage 2\n",
      "Checking ARCHIVE\n",
      "Checking BS21001 Statistics & Experimental Design Module Specification Spec QASv4.docx\n",
      "processing BS21001 Statistics & Experimental Design Module Specification Spec QASv4.docx\n",
      "Checking BS21002 Cellular Communication QASv7.docx\n",
      "processing BS21002 Cellular Communication QASv7.docx\n",
      "Checking BS21008 Organic Chemistry for the Life Sciences Spec QASv7.docx\n",
      "processing BS21008 Organic Chemistry for the Life Sciences Spec QASv7.docx\n",
      "Checking BS21010 Introductory Programming for Life Sciences Spec QASv5.docx\n",
      "processing BS21010 Introductory Programming for Life Sciences Spec QASv5.docx\n",
      "Checking BS21011 Developing Scientific Enterprise Module Specification QASv4.docx\n",
      "processing BS21011 Developing Scientific Enterprise Module Specification QASv4.docx\n",
      "Checking BS21012 Core Skills in the Life Sciences 2A QASv2.docx\n",
      "processing BS21012 Core Skills in the Life Sciences 2A QASv2.docx\n",
      "Checking BS22001 Human Physiology & Pharmacology Spec QASv5.docx\n",
      "processing BS22001 Human Physiology & Pharmacology Spec QASv5.docx\n",
      "Checking BS22002 Biomolecular Mechanisms Specification QASv5.docx\n",
      "processing BS22002 Biomolecular Mechanisms Specification QASv5.docx\n",
      "Checking BS22003 Core Skills in the Life Sciences 2B QASv4.docx\n",
      "processing BS22003 Core Skills in the Life Sciences 2B QASv4.docx\n",
      "Checking Withdrawn\n",
      "Looking in Stage 3\n",
      "Checking ARCHIVE\n",
      "Checking BS30007 Life Sciences Placement Module Specification.docx\n",
      "processing BS30007 Life Sciences Placement Module Specification.docx\n",
      "Checking BS31003 Molecular Structure & Interactions Module Specification QASv6.docx\n",
      "processing BS31003 Molecular Structure & Interactions Module Specification QASv6.docx\n",
      "Checking BS31004 Biochemistry and Cell Biology Spec QASv8.docx\n",
      "processing BS31004 Biochemistry and Cell Biology Spec QASv8.docx\n",
      "Checking BS31005 Genetics Module Specification QASv6.docx\n",
      "processing BS31005 Genetics Module Specification QASv6.docx\n",
      "Checking BS31006 Gene Regulation & Expression QASv6.docx\n",
      "processing BS31006 Gene Regulation & Expression QASv6.docx\n",
      "Checking BS31013 Biomembranes Module Specification QASv5.docx\n",
      "processing BS31013 Biomembranes Module Specification QASv5.docx\n",
      "Checking BS31016 Practical Techniques in Biomedical Sciences Spec QASv6.docx\n",
      "processing BS31016 Practical Techniques in Biomedical Sciences Spec QASv6.docx\n",
      "Checking BS31019 Regulatory Physiology and Pharmacology Module Specification QASv7.docx\n",
      "processing BS31019 Regulatory Physiology and Pharmacology Module Specification QASv7.docx\n",
      "Checking BS31020 Experimental Cell Culture Mod Spec QASv5.docx\n",
      "processing BS31020 Experimental Cell Culture Mod Spec QASv5.docx\n",
      "Checking BS31701 Human Ageing Module Specification QASv1.docx\n",
      "processing BS31701 Human Ageing Module Specification QASv1.docx\n",
      "Checking BS31702 Human Neuroanatomy Module Specification QASv1.docx\n",
      "processing BS31702 Human Neuroanatomy Module Specification QASv1.docx\n",
      "Checking BS32003 Drug Discovery and Development QASv6.docx\n",
      "processing BS32003 Drug Discovery and Development QASv6.docx\n",
      "Checking BS32004 Molecular Biology QASv7.docx\n",
      "processing BS32004 Molecular Biology QASv7.docx\n",
      "Checking BS32005 Cell and Developmental Biology QASv5.docx\n",
      "processing BS32005 Cell and Developmental Biology QASv5.docx\n",
      "Checking BS32006 Cell Signalling QASv6.docx\n",
      "processing BS32006 Cell Signalling QASv6.docx\n",
      "Checking BS32007 Organic Synthesis Specification QASv4.docx\n",
      "processing BS32007 Organic Synthesis Specification QASv4.docx\n",
      "Checking BS32008 Plant Science Specification QASv3.docx\n",
      "processing BS32008 Plant Science Specification QASv3.docx\n",
      "Checking BS32009 Immunology Specification QASv4.docx\n",
      "processing BS32009 Immunology Specification QASv4.docx\n",
      "Checking BS32010 Bioinformatics QASv4.docx\n",
      "processing BS32010 Bioinformatics QASv4.docx\n",
      "Checking BS32011 Practical Project A Module Spec QASv7.docx\n",
      "processing BS32011 Practical Project A Module Spec QASv7.docx\n",
      "Checking BS32012 Practical Project B Module Spec QASv7.docx\n",
      "processing BS32012 Practical Project B Module Spec QASv7.docx\n",
      "Checking BS32020 Human Epithelial Biology QASv3.docx\n",
      "processing BS32020 Human Epithelial Biology QASv3.docx\n",
      "Checking BS32021 Quantitative Pharmacology Module Spec QASv4.docx\n",
      "processing BS32021 Quantitative Pharmacology Module Spec QASv4.docx\n",
      "Checking BS32024 Behavioural Neuroscience and Pharmacology QASv6.docx\n",
      "processing BS32024 Behavioural Neuroscience and Pharmacology QASv6.docx\n",
      "Checking BS32026 Sensory and Motor Neuroscience QASv5.docx\n",
      "processing BS32026 Sensory and Motor Neuroscience QASv5.docx\n",
      "Checking BS32028 Molecular Pharmacology QASv3.docx\n",
      "processing BS32028 Molecular Pharmacology QASv3.docx\n",
      "Checking BS32029 Cell Proliferation Module Specification QASv2.docx\n",
      "processing BS32029 Cell Proliferation Module Specification QASv2.docx\n",
      "Checking BS32030 Data and Statistical Analysis Mod Spec QASv6.docx\n",
      "processing BS32030 Data and Statistical Analysis Mod Spec QASv6.docx\n",
      "Checking BS32031 Bioinformatics Research Skills 1 Specification QASv2.docx\n",
      "processing BS32031 Bioinformatics Research Skills 1 Specification QASv2.docx\n",
      "Checking BS32032 Science Communication Module Specification QASv4.docx\n",
      "processing BS32032 Science Communication Module Specification QASv4.docx\n",
      "Checking BS32033 Epidemiology and Health Informatics QASv1.docx\n",
      "processing BS32033 Epidemiology and Health Informatics QASv1.docx\n",
      "Checking BS32034 Introduction to Biobusiness Module Specification QASv1.docx\n",
      "processing BS32034 Introduction to Biobusiness Module Specification QASv1.docx\n",
      "Checking BS32035 Project Management and Bioenterprise QASv1.docx\n",
      "processing BS32035 Project Management and Bioenterprise QASv1.docx\n",
      "Checking BS32701 Fundamental Pharmacology Module Specification QASv1.docx\n",
      "processing BS32701 Fundamental Pharmacology Module Specification QASv1.docx\n",
      "Checking BS32702 Human Physiology Module Specification QASv1.docx\n",
      "processing BS32702 Human Physiology Module Specification QASv1.docx\n",
      "Checking BS32703 Human Physiology Module Specification QASv1.docx\n",
      "processing BS32703 Human Physiology Module Specification QASv1.docx\n",
      "Checking BS32704 Neuronal Development & Diseases Module Specification QASv1.docx\n",
      "processing BS32704 Neuronal Development & Diseases Module Specification QASv1.docx\n",
      "Checking BS32705 Neuropharmacology Module Specification QASv1.docx\n",
      "processing BS32705 Neuropharmacology Module Specification QASv1.docx\n",
      "Checking BS32706 Molecular Basis of Human Disease Module Specification QASv1.docx\n",
      "processing BS32706 Molecular Basis of Human Disease Module Specification QASv1.docx\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking BS32707 Molecular Microbiology in Human Disease Module Specification QASv1.docx\n",
      "processing BS32707 Molecular Microbiology in Human Disease Module Specification QASv1.docx\n",
      "Checking BS32708 Molecular Biophysics Module Specification QASv1.docx\n",
      "processing BS32708 Molecular Biophysics Module Specification QASv1.docx\n",
      "Looking in Stage 4\n",
      "Checking ARCHIVE\n",
      "Checking BS41004 Research Project Biological Sciences Module Specification QASv4.docx\n",
      "processing BS41004 Research Project Biological Sciences Module Specification QASv4.docx\n",
      "Checking BS41005 Research Project Biological Sciences Programme Specification QASv3.docx\n",
      "processing BS41005 Research Project Biological Sciences Programme Specification QASv3.docx\n",
      "Checking BS41006 Research Skill in Biomedical Sciences QASv5.docx\n",
      "processing BS41006 Research Skill in Biomedical Sciences QASv5.docx\n",
      "Checking BS41007 Research Project Biomedical Sciences Module Specification QASv6.docx\n",
      "processing BS41007 Research Project Biomedical Sciences Module Specification QASv6.docx\n",
      "Checking BS41009 Internship Work Placement QASv1.docx\n",
      "processing BS41009 Internship Work Placement QASv1.docx\n",
      "Checking BS41701 Honours Project in Life Sciences Module Specification QASv1.docx\n",
      "processing BS41701 Honours Project in Life Sciences Module Specification QASv1.docx\n",
      "Checking BS41702 System Neurobiology Module Specification QASv1.docx\n",
      "processing BS41702 System Neurobiology Module Specification QASv1.docx\n",
      "Checking BS41703 Functional Genomics Module Specification QASv1.docx\n",
      "processing BS41703 Functional Genomics Module Specification QASv1.docx\n",
      "Checking BS41704 Epigenetics & Chromatin Biology Module Specification QASv1.docx\n",
      "processing BS41704 Epigenetics & Chromatin Biology Module Specification QASv1.docx\n",
      "Checking BS41705 Pharmacogenetics & Drug Responses Module Specification QASv1.docx\n",
      "processing BS41705 Pharmacogenetics & Drug Responses Module Specification QASv1.docx\n",
      "Checking BS41706 Drug Discovery & Clinical Trials Module Specification QASv1.docx\n",
      "processing BS41706 Drug Discovery & Clinical Trials Module Specification QASv1.docx\n",
      "Checking BS42003 Advanced Bioinformatics QASv6.docx\n",
      "processing BS42003 Advanced Bioinformatics QASv6.docx\n",
      "Checking BS42004 Advanced Modern Drug Discovery Module Specification QASv2.docx\n",
      "processing BS42004 Advanced Modern Drug Discovery Module Specification QASv2.docx\n",
      "Checking BS42005 Advanced Plant Sciences QASv5.docx\n",
      "processing BS42005 Advanced Plant Sciences QASv5.docx\n",
      "Checking BS42006 Advanced Immunology QASv5.docx\n",
      "processing BS42006 Advanced Immunology QASv5.docx\n",
      "Checking BS42007 Cancer Biology QASv5.docx\n",
      "processing BS42007 Cancer Biology QASv5.docx\n",
      "Checking BS42008 Stem Cells in Development & Disease Module Specification QASv3.docx\n",
      "processing BS42008 Stem Cells in Development & Disease Module Specification QASv3.docx\n",
      "Checking BS42009 Advanced Molecular Microbiology QASv5.docx\n",
      "processing BS42009 Advanced Molecular Microbiology QASv5.docx\n",
      "Checking BS42010 Advanced Gene Regulation and Expression QASv5.docx\n",
      "processing BS42010 Advanced Gene Regulation and Expression QASv5.docx\n",
      "Checking BS42011 Biological Chemistry Specification QASv4.docx\n",
      "processing BS42011 Biological Chemistry Specification QASv4.docx\n",
      "Checking BS42012 Parasitology Module Spec QASv4.docx\n",
      "processing BS42012 Parasitology Module Spec QASv4.docx\n",
      "Checking BS42013 Advanced Cell Signalling QASv5.docx\n",
      "processing BS42013 Advanced Cell Signalling QASv5.docx\n",
      "Checking BS42014 Nutrients and Metabolic Disease QASv7.docx\n",
      "processing BS42014 Nutrients and Metabolic Disease QASv7.docx\n",
      "Checking BS42017 Translational Psychiatry & Neurology Module Specification QASv4.docx\n",
      "processing BS42017 Translational Psychiatry & Neurology Module Specification QASv4.docx\n",
      "Checking BS42018 Mapping Memories in the Brain QASv6.docx\n",
      "processing BS42018 Mapping Memories in the Brain QASv6.docx\n",
      "Checking BS42019 Cardiovascular Pharmacology Modu Spec QASv5.docx\n",
      "processing BS42019 Cardiovascular Pharmacology Modu Spec QASv5.docx\n",
      "Checking BS42021 Heart & Circulation Module Specification QASv6.docx\n",
      "processing BS42021 Heart & Circulation Module Specification QASv6.docx\n",
      "Checking BS42023 Applied Neuroanatomy & Neurodegenerative Disorders Module Specification QASv2.docx\n",
      "processing BS42023 Applied Neuroanatomy & Neurodegenerative Disorders Module Specification QASv2.docx\n",
      "Checking BS42025 The Conscious Brain QASv5.docx\n",
      "processing BS42025 The Conscious Brain QASv5.docx\n",
      "Checking BS42027 Cancer Pharmacology & Treatment Specification QASv5.docx\n",
      "processing BS42027 Cancer Pharmacology & Treatment Specification QASv5.docx\n",
      "Checking BS42028 Pharmacology of Treatment of Metabolic Disease Module Specification QASv2.docx\n",
      "processing BS42028 Pharmacology of Treatment of Metabolic Disease Module Specification QASv2.docx\n",
      "Checking BS42701 Extreme Physiology Module Specification QASv1.docx\n",
      "processing BS42701 Extreme Physiology Module Specification QASv1.docx\n",
      "Checking BS42702 Functional Ageing Module Specification QASv1.docx\n",
      "processing BS42702 Functional Ageing Module Specification QASv1.docx\n",
      "Checking BS42703 Genetic Medicine in the Post-Genomic Era Module Specification QASv1.docx\n",
      "processing BS42703 Genetic Medicine in the Post-Genomic Era Module Specification QASv1.docx\n",
      "Checking BS42704 Advanced Cell Biology Module Specification QASv1.docx\n",
      "processing BS42704 Advanced Cell Biology Module Specification QASv1.docx\n",
      "Checking Withdrawn\n",
      "Looking in Stage 5\n",
      "Checking BS51007 Cell Biology and Disease Module Specification QASv3.docx\n",
      "processing BS51007 Cell Biology and Disease Module Specification QASv3.docx\n",
      "Checking BS51008 Molecular Basis of Disease QASv2.docx\n",
      "processing BS51008 Molecular Basis of Disease QASv2.docx\n",
      "Checking BS51009 Scientific Method and Skills Module Specification QASv4.docx\n",
      "processing BS51009 Scientific Method and Skills Module Specification QASv4.docx\n",
      "Checking BS51010 Scientific and Health Landscapes Module Spec QASv3.docx\n",
      "processing BS51010 Scientific and Health Landscapes Module Spec QASv3.docx\n",
      "Checking BS51011 Writing and Communication Module Spec QASv4.docx\n",
      "processing BS51011 Writing and Communication Module Spec QASv4.docx\n",
      "Checking BS51012 Tools Software and Databases Module Spec QASv3.docx\n",
      "processing BS51012 Tools Software and Databases Module Spec QASv3.docx\n",
      "Checking BS52007 Advanced Techology Platfrorms QASv3.docx\n",
      "processing BS52007 Advanced Techology Platfrorms QASv3.docx\n",
      "Checking BS52008 MSc Project Proposal and Associated Skills Specification QASv3.docx\n",
      "processing BS52008 MSc Project Proposal and Associated Skills Specification QASv3.docx\n",
      "Checking BS52009 Therapeutics Specification QASv2.docx\n",
      "processing BS52009 Therapeutics Specification QASv2.docx\n",
      "Checking BS52010 Communication and Marketing Strategy QASv3.docx\n",
      "processing BS52010 Communication and Marketing Strategy QASv3.docx\n",
      "Checking BS52011 Sci & Health Comm Industries and Contemporary Issues Module Spec QASv3.docx\n",
      "processing BS52011 Sci & Health Comm Industries and Contemporary Issues Module Spec QASv3.docx\n",
      "Checking BS52012 Digital and Visual Communication QASv3.docx\n",
      "processing BS52012 Digital and Visual Communication QASv3.docx\n",
      "Checking BS52013 Biomolecular Modelling Module Spec QASv1.docx\n",
      "processing BS52013 Biomolecular Modelling Module Spec QASv1.docx\n",
      "Checking BS53003 MSc Biobusiness Research Project Spec QASv4.docx\n",
      "processing BS53003 MSc Biobusiness Research Project Spec QASv4.docx\n",
      "Checking BS53004 MSc Research Project QASv4.docx\n",
      "processing BS53004 MSc Research Project QASv4.docx\n",
      "Checking BS53005 Science and Health Communications Masters Project Module Spec QASv3.docx\n",
      "processing BS53005 Science and Health Communications Masters Project Module Spec QASv3.docx\n",
      "Checking BS53006 Masters Placement Module Spec QASv2.docx\n",
      "processing BS53006 Masters Placement Module Spec QASv2.docx\n",
      "Checking BS53007 MSc Quantitative Biology Research Project QASv1.docx\n",
      "processing BS53007 MSc Quantitative Biology Research Project QASv1.docx\n"
     ]
    }
   ],
   "source": [
    "modules = {}\n",
    "for stage in os.listdir(moduledir):\n",
    "    print(f'Looking in {stage}')\n",
    "    for file in os.listdir(os.path.join(moduledir, stage)):\n",
    "        print(f'Checking {file}')\n",
    "        if file.endswith('.docx'):\n",
    "            print(f'processing {file}')\n",
    "            doc=docx.Document(os.path.join(moduledir, stage, file))\n",
    "            structure ={}\n",
    "            for t in doc.tables:\n",
    "                for r in range(len(t.rows)):\n",
    "                    structure[t.row_cells(r)[0].text.split()[0]] =t.row_cells(r)[1].text\n",
    "            modules[structure['1.2']]=structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "aa65577a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'code': 'Module code', 'name': 'Module name', 'credits': 'Credit weight', 'scqflevel': 'SCQF level', 'shelevel': 'SHE level', 'semester': '1, 2 or both'} {'block': 'Teaching block: C or D', 'startyear': 'First academic year', 'endyear': 'Last academic year'}\n"
     ]
    }
   ],
   "source": [
    "print(Module.requiredParams, Module.optionalParams)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "ddfa345f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BS10001 1.14 \n",
      "BS11008 1.14 \n",
      "BS11009 1.14 \n",
      "BS12009 1.14 \n",
      "BS12010 1.14 \n",
      "BS21002 1.14 \n",
      "BS21012 1.14 \n",
      "BS30007 1.14 \n",
      "BS32034 1.14 \n",
      "BS41009 1.14 \n",
      "BS42028 1.14 \n",
      "BS51007 1.14 \n",
      "BS51008 1.14 \n",
      "BS51009 1.14 \n",
      "BS51010 1.14 \n",
      "BS51011 1.14 \n",
      "BS51012 1.14  \n",
      "BS52007 1.14 \n",
      "BS52009 1.14 \n",
      "BS52010 1.14 \n",
      "BS52011 1.14 \n",
      "BS52012 1.14 \n",
      "BS52013 1.14 Once per academic year in semester 2\n",
      "BS53003 1.14 \n",
      "BS53004 1.14 \n",
      "BS53005 1.14  semester 3 (summer) \n",
      "BS53006 1.14  semester 3 (summer) \n",
      "BS53007 1.14 \n"
     ]
    }
   ],
   "source": [
    "newmod = modules['BS32031']\n",
    "modobj ={}\n",
    "\n",
    "scqf = re.compile(r'SCQF\\D+(\\d+)\\D+(\\d+)\\D')\n",
    "delivery = re.compile(r'Semester (\\d)')\n",
    "#for m in modules:\n",
    "#    print(m)\n",
    "#for m in ['BS32031']:\n",
    "for m in modules:\n",
    "    try:\n",
    "        newmod=modules[m]\n",
    "    \n",
    "        match=scqf.search(newmod['1.3'])\n",
    "        scqflevel=0\n",
    "        shelevel=0\n",
    "        credits=0\n",
    "        if match:\n",
    "            scqflevel = int(match.group(1))\n",
    "            shelevel = scqflevel - 6\n",
    "            credits = int(match.group(2))\n",
    "        else: \n",
    "            shelevel = int(m[2])\n",
    "            scqflevel = shelevel +6\n",
    "            print(m,\"1.3\", newmod['1.3'])\n",
    "        match=delivery.search(newmod['1.14'])\n",
    "        semester=0\n",
    "        if match:\n",
    "            semester = int(match.group(1))\n",
    "        else:\n",
    "            print(m,\"1.14\",newmod['1.14'])\n",
    "            semester = int(m[3])\n",
    "        if credits and scqflevel and shelevel and semester:\n",
    "            newmodparams ={'code':newmod['1.2'], 'name':newmod['1.1'], 'semester':semester, \n",
    "                           'credits': credits, 'scqflevel':scqflevel, 'shelevel':shelevel}\n",
    "            modobj[m]=factory.get_or_create_Element('Module', **newmodparams)\n",
    "    except Exception as e:\n",
    "        print(newmodparams,e)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "582ed72c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BS11005: Prerequisites: After completion of the skills audit prior to matriculation, a decision on whether this module is to be taken by the student will be made\n",
      "BS11005: Corequisites: For students enrolled for Biological and Biomedical Sciences degrees, the following are all co-requisites:\n",
      "   BS11009 (knowledge based)\n",
      "   BS11008 (Lab and skills-based)\n",
      "\n",
      "BS11008: Corequisites: For students enrolled for Biological Sciences and Biomedical Sciences degrees, the following module is a co-requisites:\n",
      "• BS11009 (Knowledge based)\n",
      "BS12010: Corequisites: For students enrolled for Biological Sciences and Biomedical Sciences degrees, the following modules are all co-requisites:\n",
      "• BS12001 (Knowledge based)\n",
      "• BS12002 (Knowledge based)\n",
      "BS12701: Prerequisites: Successful completion of JPD Core curriculum to date\n",
      "BS12702: Prerequisites: Successful completion of JPD Core curriculum to date\n",
      "BS21001: Prerequisites: BS11001, BS11002, BS11008, BS12001, BS12002, BS12010 or Level 2 direct entry requirements\n",
      "\n",
      "BS21002: Prerequisites: BS11009, BS11008, BS12011,\n",
      "BS12010 or Level 2 direct entry requirements\n",
      "BS21008: Prerequisites: BS11009, BS11008, BS12011, BS12010 or their equivalent.  Prerequisite module BS12009.\n",
      "BS21008: Antirequisites: BS12006\n",
      "BS21010: Antirequisites: EG21007\n",
      "BS21012: Corequisites: • BS21001 (Knowledge based)\n",
      "• BS21002 (Knowledge based)\n",
      "BS22001: Prerequisites: BS11001, BS11002, BS11003, BS11004, BS12001,\n",
      "BS12002 BS12003, BS12004 or Level 2 direct entry requirements\n",
      "\n",
      "BS22001: Corequisites: BS21001, BS21002, BS21003, BS22002 and BS22003\n",
      "BS22002: Prerequisites: BS11001, BS11002, BS11008, BS12001, BS12002, BS12010 or have experience of basic Life Sciences laboratory techniques\n",
      "\n",
      "BS22003: Prerequisites: BS11001, BS11002,BS11008, BS12001, BS12002, BS12010 or have experience of basic Life Sciences laboratory techniques\n",
      "\n",
      "BS22003: Corequisites: BS22001\n",
      "BS30007: Prerequisites: Must have successfully completed 360 SCQF credits, including 120 at level 9, in any of the following programmes:\n",
      "BSc (Hons) Biological Sciences\n",
      "BSc (Hons) Biological Sciences (Bioinformatics)\n",
      "BSc (Hons) Biological Sciences (Plant Sciences)\n",
      "BSc (Hons) Biochemistry\n",
      "BSc (Hons) Microbiology\n",
      "BSc (Hons) Biological Chemistry and Drug Discovery\n",
      "BSc (Hons) Biomedical Sciences\n",
      "BSc (Hons) Neuroscience\n",
      "BSc (Hons) Pharmacology\n",
      "BS31003: Prerequisites: Successful completion of Level 1&2 Life Sciences Core Curriculum or similar experience of basic Life Science education.\n",
      "BS31003: Corequisites: Completion of three additional (15 credit) Level 3\n",
      "modules in Semester 1 two of which must be from the Biological Sciences Stream\n",
      "In conjunction with an appropriate combination of Level 3 modules from Semester 1 which are relevant to the students’ chosen degree programme.\n",
      "\n",
      "BS31004: Prerequisites: Successful completion of Level 2 Biological/Biomedical Sciences or Equivalent\n",
      "BS31005: Prerequisites: Successful completion of Level 1&2 Life Sciences Core Curriculum or similar experience of basic Life Science education.\n",
      "BS31006: Prerequisites: Successful completion of Level 1 & 2 Life Sciences Core Curriculum or similar experience of basic Life Sciences education.\n",
      "BS31006: Corequisites: Completion of three additional (15 credit) Level 3 modules in Semester 1 which are relevant to the students' chosen degree programme.\n",
      "BS31013: Prerequisites: Successful completion of Level 1&2 Life Sciences Core Curriculum or similar experience of basic Life Sciences Education\n",
      "BS31016: Prerequisites: Successful completion of Level 1 & 2 Life Sciences Core Curriculum or similar experience of basic Life Science education.\n",
      "This module will only be available to Level 3 Biomedical Stream Students.\n",
      "\n",
      "BS31019: Prerequisites: Students will be expected have completed the following core modules at Level 2 or have completed modules of similar content from partner institutions from which exchange students are joining the Dundee taught module (BS21001, Statistics and Experimental Design; BS21002, The Gene and the Cell; BS21003, Laboratory and Research Skills 2A; BS21004, Laboratory and Research Skills 2B; BS22001, Biomedical Sciences; BS22002, Biological Sciences; BS22003, Laboratory and Research Skills).\n",
      "BS31020: Prerequisites: Successful completion of Level 1 & 2 Biological and Biomedical Sciences Core Curriculum or similar experience of basic Life Science education.\n",
      "This module will only be available to Level 3 Students from the Biomedical stream.\n",
      "\n",
      "BS31701: Prerequisites: Successful completion of JPD Core curriculum to date\n",
      "BS31702: Prerequisites: Successful completion of JPD Core curriculum to date\n",
      "BS32003: Prerequisites: Successful completion of Level 1&2 Life Sciences Core Curriculum and semester 1 of the Level 3 Biological Sciences stream or similar experience of basic Life Sciences Education.\n",
      "BS32003: Corequisites: BS32007 Organic Synthesis and two additional (15 credit) Level 3 modules in Semester 2, one of which must be from the Biological Sciences Stream.\n",
      "BS32004: Prerequisites: Successful completion of Level 1&2 Life Sciences Core Curriculum or similar experience of basic Life Science education. Study of 60 Level 3 Credits in Semester 1.\n",
      "BS32005: Prerequisites: Successful completion of Level 1&2 and Level 3 Sem 1 Life Sciences Core modules BS31004 and/or BS31006 or similar experience of basic Life Science education.\n",
      "BS32006: Prerequisites: Successful completion of Level 1&2 Biological and Biomedical Core Curriculum or similar experience of basic Life Science education. Study of 60 Level 3 Credits in Semester 1.\n",
      "BS32007: Prerequisites: Successful completion of 120 credits at Level 2.\n",
      "BS32008: Prerequisites: Successful completion of Level 1&2 Biological and Biomedical Sciences Core Curriculum or similar experience of basic Life Science education. Study of 60 Level 3 Credits in Semester 1.\n",
      "BS32009: Prerequisites: Successful completion of Level 1&2 Biological and Biomedical Sciences Core Curriculum or similar experience of basic Life Science education. Study of 60 Level 3 Credits or equivalent in Semester 1.\n",
      "BS32010: Prerequisites: Successful completion of Level 1 and 2 Biological and Biomedical Sciences core curriculum or similar experience of Life Science education.\n",
      "BS32011: Prerequisites: Successful completion of Level 1&2 Biological and Biomedical Sciences Core Curriculum or similar experience of basic Life Science education. Study of 60 Level 3 Credits in Semester 1 or an appropriately mapped learning experience.\n",
      "BS32011: Corequisites: This module may be matched to a Semester 2 specialist theory module.\n",
      "BS32012: Prerequisites: Successful completion of Level 1&2 Biological and Biomedical Sciences Core Curriculum or similar experience of basic Life Science education. Study of 60 Level 3 Credits in Semester 1 or an appropriately mapped learning experience.\n",
      "BS32012: Corequisites: This module may be matched to a Semester 2 specialist theory module.\n",
      "BS32020: Prerequisites: Successful completion of Level 1&2 Biological and Biomedical Sciences Core Curriculum or similar experience of basic Life Sciences Education\n",
      "BS32021: Prerequisites: Successful completion of Level 1 & 2 Biological and Biomedical Sciences Core Curriculum or similar experience of basic Life Sciences Education\n",
      "BS32024: Prerequisites: Successful completion of Level 1&2 Biological and Biomedical Sciences Core Curriculum or similar experience of basic Life Sciences Education\n",
      "BS32026: Prerequisites: Successful completion of Level 1&2 Biological and Biomedical Sciences Core Curriculum or similar experience of basic Life Sciences Education\n",
      "BS32028: Prerequisites: Successful completion of Level 1&2 Biological and Biomedical Sciences Core Curriculum or similar experience of basic Life Sciences Education\n",
      "BS32029: Prerequisites: Successful completion of Level 1&2 Biological and Biomedical Sciences Core Curriculum or similar experience of basic Life Sciences Education \n",
      "BS32030: Prerequisites: Successful completion of Level 1&2 Biological and Biomedical Sciences Core Curriculum or similar experience of basic Life Science education.\n",
      "BS32031: Antirequisites:  BS21010, EN21007\n",
      "BS32032: Prerequisites: Successful completion of Level 1&2 Biological and Biomedical Sciences Core Curriculum or similar experience of basic Life Science education. Study of 60 Level 3 Credits in Semester 1.\n",
      "BS32032: Antirequisites: Prior completion of Level 4 Science Communication Honours project in BS41004 or BS41007\n",
      "BS32033: Prerequisites: Successful completion of Level 1&2 Biological and Biomedical Sciences Core Curriculum or similar experience of basic Life Sciences Education\n",
      "BS32035: Prerequisites: Successful completion of year 2 at HAN University of Applied Sciences  Arnhem and Nijmegen\n",
      "BS32701: Prerequisites: Successful completion of JPD Core curriculum to date\n",
      "BS32702: Prerequisites: Successful completion of JPD Core curriculum to date\n",
      "BS32703: Prerequisites: Successful completion of JPD Core curriculum to date\n",
      "BS32703: Antirequisites: LSM3212\n",
      "BS32704: Prerequisites: Successful completion of JPD Core curriculum to date\n",
      "BS32704: Antirequisites: LSM3213\n",
      "BS32705: Prerequisites: Successful completion of JPD Core curriculum to date\n",
      "BS32705: Antirequisites: LSM3221\n",
      "BS32706: Prerequisites: Successful completion of JPD Core curriculum to date\n",
      "BS32707: Prerequisites: Successful completion of JPD Core curriculum to date\n",
      "BS32708: Prerequisites: Successful completion of JPD Core curriculum to date\n",
      "BS41004: Prerequisites: Successful completion of 120 credits at Level 3 or through an intercalated BMSc programme.\n",
      "BS41004: Corequisites: Research Skills Module BS41005\n",
      "BS41005: Prerequisites: Successful completion of 120 credits at Level 3 or through an intercalated BMSc programme\n",
      "BS41005: Corequisites: BS41004 Research Project Biological Sciences\n",
      "BS41006: Prerequisites: Successful completion of 120 credits at Level 3 or through an intercalated BMSc programme\n",
      "BS41006: Corequisites: BS41007 Research Project Biomedical Sciences\n",
      "BS41007: Prerequisites: Successful completion of 120 credits at Level 3 or through an intercalated BMSc programme.\n",
      "BS41007: Corequisites: Research Skills Module BS41006\n",
      "BS41009: Prerequisites: Successful completion of Level 3\n",
      "BS41701: Prerequisites: Successful completion of JPD Core curriculum to date\n",
      "BS41702: Prerequisites: Successful completion of JPD Core curriculum to date and LSM3216\n",
      "BS41703: Prerequisites: Successful completion of JPD Core curriculum to date\n",
      "BS41704: Prerequisites: Successful completion of JPD Core curriculum to date\n",
      "BS41705: Prerequisites: Successful completion of JPD Core curriculum to date and either LSM3211 or LSM3212\n",
      "BS41706: Prerequisites: Successful completion of JPD Core curriculum to date and LSM3216\n",
      "BS42003: Prerequisites: Successful completion of 120 credits at level 3. Students from other Schools may be eligible to take this module subject to agreement.\n",
      "BS42004: Prerequisites: Successful completion of 120 credits at Level 3. Successful completion of BS32003 Drug Discovery and Development, and BS32007 Organic Chemistry\n",
      "BS42005: Prerequisites: Successful completion of 120 credits at level 3.\n",
      "BS42006: Prerequisites: Successful completion of 120 credits at Level 3 including BS32009 Immunology, and BS32006 Cell Signalling or BS32028 Molecular Pharmacology.\n",
      "BS42007: Prerequisites: Successful completion of 120 credits at Level 3 Modules including BS32029, Cell Proliferation and Survival Mechanisms underlying disease in combination with one of BS31006 Gene Regulation and Expression, BS31004 Biochemistry and Cell Biology, BS31005 Genetics, or BS31020 Experimental Cell Culture \n",
      "BS42008: Prerequisites: Successful completion of 120 credits at Level 3.\n",
      "BS42010: Prerequisites: Successful completion of 120 credits at Level 3. Three Level 3 Modules:  BS31004 Biochemistry & Cell Biology,  BS31005 Genetics,  BS31006  Gene Regulation & Expression\n",
      "BS42011: Prerequisites: Successful completion of 120 credits at Level 3 including BS32007 Organic Synthesis.\n",
      "BS42012: Prerequisites: Successful completion of 120 credits at Level 3.\n",
      "BS42013: Prerequisites: Successful completion of 120 credits at Level 3, including Cell Signalling (BS32006)\n",
      "BS42014: Prerequisites: Students will be expected have completed the following core modules at Level 2 or have completed modules of similar content from partner institutions if part of an exchange programme:  BS21001 (The Evolution of Modern Life); BS21002 (The Gene and the Cell); BS21003 (Laboratory and Research Skills 2A); BS21004 (Laboratory and Research Skills 2B); BS22001 (Biomedical Sciences); BS22002 (Biological Sciences); BS22003 (Laboratory and Research Skills).  At Level 3 students will have been expected to have completed BS31013 (Biomembranes); BS31016 (Practical Techniques in Biomedical Sciences) and BS31019 (Regulatory Physiology & Pharmacology).  Students who may have completed BS31003 (Biochemistry and Cell Biology) and BS31006 (Gene Regulation and Expression) instead of the indicated Level 3 modules would also be considered for this Level 4 module.\n",
      "BS42017: Prerequisites: Successful completion of 120 credits at level 3.\n",
      "BS42018: Prerequisites: Successful completion of 120 credits at Level 3 which ideally should include BS32024 Neuropsychopharmacology.\n",
      "BS42019: Prerequisites: Successful completion of 120 credits at Level 3 and completion of BS42021 Heart and Circulation\n",
      "BS42019: Corequisites: BS42021 in Block C\n",
      "BS42021: Prerequisites: Successful completion of 120 credits at Level 3\n",
      "BS42023: Prerequisites: Successful completion of 120 credits at Level 3.\n",
      "BS42025: Prerequisites: Successful completion of 120 credits at Level 3\n",
      "BS42027: Prerequisites: Successful completion of BS42007 120 credits at Level 3 Modules including BS32029, Cell Proliferation and Survival Mechanisms underlying disease in combination with one of BS31006 Gene Regulation and Expression, BS31004 Biochemistry and Cell Biology, BS31020 Experimental Cell Culture or BS31005 Genetics.\n",
      "BS42027: Corequisites: BS42007 in Block C\n",
      "BS42028: Prerequisites: Successful completion of Levels 1, 2 and 3 Life Sciences programmes. Successful completion of level 4 Semester 1. \n",
      "BS42028: Corequisites: BS42014 in Block D\n",
      "BS42701: Prerequisites: Successful completion of JPD Core curriculum to date and LSM3212 and LSM3214\n",
      "BS42702: Prerequisites: Successful completion of JPD Core curriculum to date and either LSM3217 or LSM3224\n",
      "BS42703: Prerequisites: Successful completion of JPD Core curriculum to date\n",
      "BS42704: Prerequisites: Successful completion of JPD Core curriculum to date\n",
      "BS51007: Prerequisites: Successful completion of the relevant semester 1 MSc modules totalling 60 credits\n",
      "BS51007: Corequisites: BS51009 Scientific Methods and Skills and either BS51008 The Molecular Basis of Disease: Genetics, Epigenetics and Genomics or\n",
      "BU52033 Leadership and Decision Making, or\n",
      "BU52031 Strategic Management & Organisational Analysis, or\n",
      "BU51017 Customer Experience Design, or \n",
      "BU51015 Principles of Marketing as Creative Practice\n",
      "BS51008: Prerequisites: Successful completion of the relevant semester 1 MSc modules totalling 60 credits\n",
      "BS51008: Corequisites: BS510xx Scientific Method and Skills and either BS510xx Cell Biology and Diseases, or\n",
      "BU52033 Leadership and Decision Making, or\n",
      "BU52031 Strategic Management & Organisational Analysis, or\n",
      "BU51017 Customer Experience Design, or \n",
      "BU51015 Principles of Marketing as Creative Practice\n",
      "BS51009: Prerequisites:  Entry onto relevant Masters course\n",
      "BS51009: Corequisites: Co-requisites are explained in the programme specifications for the MSc Biomedical and Molecular Sciences (with Business/Marketing/Management/Entrepreneurship) or MSc Applied Neuroscience or MFA Art, Science and Visual Thinking\n",
      "BS51010: Prerequisites: Entry onto relevant Masters course\n",
      "BS51011: Prerequisites: Entry onto relevant Masters course\n",
      "\n",
      "BS51012: Prerequisites: Entry onto relevant Masters course\n",
      "BS52007: Prerequisites: Successful completion of the relevant semester 1 MSc modules totalling 60 credits\n",
      "BS52007: Corequisites: BS52008 Project Proposal and Associated Skills and \n",
      "and BS52009 Therapeutics\n",
      "BS52008: Prerequisites: Successful completion of the relevant semester 1 MSc modules totalling 60 credits\n",
      "BS52008: Corequisites: Co-requisites are explained in the Schedule of modules (section 2.10) of the Programme Specifications for the MSc Biomedical and Molecular Sciences (with Business/Marketing/Management/Entrepreneurship).\n",
      "BS52009: Prerequisites: Successful completion of the relevant semester 1 MSc modules totalling 60 credits\n",
      "BS52009: Corequisites: BS52008 Project Proposal and Associated Skills and \n",
      "either BS52007 Advanced Technology Platforms, or\n",
      "BU51016 International Business Strategy, or\n",
      "BU51024 Managing People and Organisations, or\n",
      "BU52025 Digital Marketing, or\n",
      "BU52040 Practicing Entrepreneurship, or\n",
      "BU52041 Understanding Entrepreneurship in a Global Context\n",
      "BS52010: Prerequisites: Entry onto relevant Masters course\n",
      "\n",
      "BS52012: Prerequisites: Entry onto relevant Masters course\n",
      "\n",
      "BS52013: Prerequisites: Successful completion of the relevant semester 1 MSc modules totalling 60 credits\n",
      "BS53003: Prerequisites: Successful completion of the relevant semester 1 and 2 MSc modules totalling 120 credits\n",
      "BS53004: Prerequisites: Successful completion of the relevant semester 1 and 2 MSc modules totalling 120 credits\n",
      "BS53005: Prerequisites: Successful completion of the relevant semester 1 and 2 MSc modules totalling 120 credits\n",
      "BS53005: Antirequisites: BS53006 Masters Placement: Science and Health Communication\n",
      "BS53006: Prerequisites:  Successful completion of the relevant semester 1 and 2 MSc modules totalling 120 credits\n",
      "BS53006: Antirequisites: Students cannot take any additional modules during the time period of this module.\n",
      "BS53XXX Science and Health Communication Masters Project\n",
      "BS53007: Prerequisites: Successful completion of the relevant semester 1 and 2 MSc modules totalling 120 credits\n"
     ]
    }
   ],
   "source": [
    "for m in modules:\n",
    "    if modules[m]['1.16'].strip() and modules[m]['1.16'].strip().upper()[0]!='N':\n",
    "        print(f'{m}: Prerequisites: {modules[m][\"1.16\"]}')\n",
    "        modules[m]['Pre']=modules[m][\"1.16\"]\n",
    "    if modules[m]['1.17'].strip() and modules[m]['1.17'].strip().upper()[0]!='N':\n",
    "        print(f'{m}: Corequisites: {modules[m][\"1.17\"]}')\n",
    "        modules[m]['Co']=modules[m][\"1.17\"]\n",
    "    if modules[m]['1.18'].strip() and modules[m]['1.18'].strip().upper()[0]!='N':\n",
    "        print(f'{m}: Antirequisites: {modules[m][\"1.18\"]}')\n",
    "        modules[m]['Anti']=modules[m][\"1.18\"]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "b92cea31",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applicability\n",
      "Section 1: Module specification\n",
      "Heading\n",
      "1.1 Module title\n",
      "1.2 Module code\n",
      "1.3 Scottish Credit and Qualifications Framework (SCQF) level and credits\n",
      "1.4 Aims\n",
      "1.5 Indicative content\n",
      "1.6 Intended learning outcomes (ILOs)\n",
      "1.7 Assessment strategy\n",
      "1.8 Assessment types and weighting\n",
      "1.9 Teaching and learning\n",
      "1.10 Indicative student contact hours and the expected amount of independent study\n",
      "1.11 Teaching staff\n",
      "1.12 Mode of study\n",
      "1.13 Location(s) for the delivery of teaching\n",
      "1.14 When delivered \n",
      "1.15 Equality and diversity\n",
      "1.16 Pre-requisites or entry requirements\n",
      "1.17 Co-requisites\n",
      "1.18 Anti-requisites\n",
      "1.19 Student numbers\n",
      "Section 2: Responsibilities and contacts\n",
      "Heading\n",
      "2.1 Lead School\n",
      "2.2 Lead discipline\n",
      "2.3 Collaborating Schools (where relevant)\n",
      "2.4 Module leader\n",
      "2.5 Lead administrator for the module\n",
      "Section 3: Module details (information for internal use for quality assurance and management)\n",
      "Heading\n",
      "3.1 Rationale for the module\n",
      "3.2 First student intake\n",
      "3.3 External organisations associated with the module\n",
      "3.4 Reference points for quality and academic standards\n",
      "3.5 Teaching quality and staff development opportunities\n",
      "3.6 Distance or blended learning\n",
      "3.7 Placements\n",
      "3.8 Inclusive nature of the curriculum and its delivery\n",
      "3.9 Teaching accommodation\n",
      "3.10 Learning and assessment resources\n",
      "3.11 Other resources including any specialist equipment or facilities\n",
      "3.12 Additional costs for students\n",
      "3.13 Teaching allocations\n",
      "Section 4: University management information (to be completed by the University’s Quality and Academic Standards office)\n",
      "Heading\n",
      "Version\n",
      "Status\n",
      "Date approved by QASC or, for minor changes, noted by the QAS office.\n"
     ]
    }
   ],
   "source": [
    "for t in doc.tables:\n",
    "    for r in range(len(t.rows)):\n",
    "                    print(t.row_cells(r)[0].text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "f3214d64",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "class Node():\n",
    "    '''Represents a Node in a Neo4j database. \n",
    "    Contains methods for accessing links, ensuring that on creation the \n",
    "    correct parameters are given etc.'''\n",
    "\n",
    "    ntype='Node'\n",
    "    \n",
    "    requiredParams = {\n",
    "        }\n",
    "    optionalParams= {}\n",
    "    \n",
    "    def __init__(self, factory=None,  **kwparams):\n",
    "        '''Initialiser. Checks required parameters are present'''\n",
    "        if factory is None:\n",
    "            raise Exception('Curriculum factory must be given as factory')\n",
    "        self.factory = factory\n",
    "        self.ntype=self.__class__.__name__.split('.')[-1]\n",
    "        self.paramChecks={'default': self._checkDefault}\n",
    "        self.params={}\n",
    "        self.edges=[]\n",
    "        self._get_labels()\n",
    "        for par in self.requiredParams:\n",
    "            if not kwparams.get(par):\n",
    "                print(par, kwparams)\n",
    "                raise InsufficientParameterSpecException(f'Missing parameter <{par}> of type <{self.requiredParams[par]}>')\n",
    "            if not self.paramChecks.get(par,self._checkDefault)(kwparams[par]):\n",
    "                raise BadParameterException(f'Parameter <{par}> should be of type <{self.requiredParams[par]}>')\n",
    "            self.params[par]=kwparams[par]\n",
    "        for par in self.optionalParams:\n",
    "            if par not in kwparams:\n",
    "                continue\n",
    "            if not self.paramChecks.get(par,self._checkDefault)(kwparams[par]):\n",
    "                raise BadParameterException(f'Parameter <{par}> should be of type <{self.requiredParams[par]}>')\n",
    "            self.params[par]=kwparams[par]\n",
    "        if 'elementID' not in kwparams:\n",
    "            self._create_node()\n",
    "        else:\n",
    "            self.element_id=kwparams['elementID']\n",
    "            self.edges =self.getEdges()\n",
    "    def __str__(self):\n",
    "        '''Returns a string represntation of the node'''\n",
    "        params = \", \".join([f'{k}: {self.params[k]}' for k in self.requiredParams])\n",
    "        return f'{self.ntype}: [{params}] ID: {self.element_id}'\n",
    "    \n",
    "    def __repr__(self):\n",
    "        return self.__str__()\n",
    "                            \n",
    "    def _create_node(self):\n",
    "        '''Internal method that creates a new instance of the node, if it does not exist'''\n",
    "        paramlist = \", \".join([f'{k}: ${k}' for k in self.params])\n",
    "        paramlist = \"{\"+paramlist+\"}\"\n",
    "        nodelabels = ':'.join(self.labels)\n",
    "        query= f\"MERGE (a:{nodelabels} {paramlist}) RETURN a\"\n",
    "        records,summary, keys = self.factory.db.execute_query(query, self.params, database_=self.factory.dbname)   \n",
    "        self.element_id = records[0].items()[0][1].element_id\n",
    "            \n",
    "    def _checkDefault(self, value):\n",
    "        '''default placeholder for parameter value checks. Returns True'''\n",
    "        return True\n",
    "    \n",
    "    def _checkAY(self, value):\n",
    "        '''Checks for a correctly formatted Academic year as:\n",
    "            2425\n",
    "            24/25\n",
    "            2024/5'''\n",
    "        #TODO \n",
    "        return True\n",
    "    def getparam(self, prop, default=None):\n",
    "        '''\n",
    "        Returns a parameter value (default if not found)\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        prop : property name\n",
    "            \n",
    "        default : TYPE, optional\n",
    "             The default is None.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        Property value\n",
    "            \n",
    "        '''\n",
    "        return self.params.get(prop,default)\n",
    "    \n",
    "    def setparam(self, key, value):\n",
    "        '''\n",
    "        Sets the parameter key to value. No type checking.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        key : legal value for a dictionary key\n",
    "            \n",
    "        value : Value to store.\n",
    "        \n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        self.params[key]=value\n",
    "        self.update()\n",
    "    \n",
    "    def getEdges(self,relation=None,**kwargs):\n",
    "        '''retrieves edges connected to or from the Node.\n",
    "        Relation is the type of relation to retrieve (default is all)\n",
    "        other arguments are applied as matching parameters'''\n",
    "        edges=[]\n",
    "        filterstring=''\n",
    "        filterparams={}\n",
    "        for kw in kwargs:\n",
    "            if kw != 'id':\n",
    "                filterstring += f\"'{kw}': ${kw},\"\n",
    "                filterparams[kw] = kwargs[kw]\n",
    "        rel=''\n",
    "        if relation:\n",
    "            rel=f\":{relation}\"\n",
    "        filterparams['id'] = self.element_id\n",
    "        if filterstring:\n",
    "            filterstring = \"{\"+filterstring+\"}\"\n",
    "        records, summary, keys = self.factory.db.execute_query(\n",
    "        f\"MATCH (p: {self.ntype} {{id: $id }}) -[e{rel} {filterstring}]-(q) RETURN e,q\",\n",
    "        filterparams,\n",
    "        routing_=neo4j.RoutingControl.READ,  # or just \"r\"\n",
    "        database_=self.factory.dbname)\n",
    "        for edge in records:\n",
    "            relation = edge.items[0][1]\n",
    "            target = edge.items[1][1]\n",
    "            edges.append({'source': self,'edge': relation, 'target': target})\n",
    "        return edges\n",
    "    def isCurrent(self, year):\n",
    "        '''\n",
    "        If the edge has a start/end year, check if the given year is the period in which the relation is valid\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        Boolean: True if the eyar is between start and end (inclusive), after start if no end, or True if no start\n",
    "\n",
    "        '''\n",
    "        \n",
    "        if self.params.get('startyear'):\n",
    "            if AcademicYear(year)< AcademicYear(self.params.get('startyear')):\n",
    "                return False\n",
    "            if self.params.get('endyear'):\n",
    "                if AcademicYear(year)> AcademicYear(self.params.get('endyear')):\n",
    "                    return False\n",
    "        return True\n",
    "    \n",
    "    \n",
    "    def toDict(self):\n",
    "        retp ={'element_id': self.element_id}\n",
    "        for p in self.params:\n",
    "            retp[p]=self.params[p]\n",
    "        return retp\n",
    "    \n",
    "    def update(self, **kwargs):\n",
    "        '''\n",
    "        Update named parameters for a Node\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        **kwargs : key,value parameter set\n",
    "            any key,value parameters. Existing values will be overwritten.\n",
    "            No type checking\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        paramlist =[]\n",
    "        if not kwargs:\n",
    "            return\n",
    "        for k in kwargs:\n",
    "            if k !='id':\n",
    "                paramlist.append( f'a.{k}= ${k}')\n",
    "                self.params[k]=kwargs[k]\n",
    "        kwargs['id'] = self.element_id    \n",
    "        paramlist = \"{\"+paramlist+\"}\"\n",
    "        query= f\"MERGE (a:{self.ntype}) WHERE elementID(a) = $id SET {', '.join(paramlist)}\"\n",
    "        records,summary, keys = self.factory.db.execute_query(query,kwargs, database_=self.factory.dbname)   \n",
    "        \n",
    "    def _get_labels(self):\n",
    "        self.labels = [x.__name__ for x in globals()[self.__class__.__name__].__mro__]\n",
    "        self.labels.remove('object')\n",
    "        self.labels.remove('Node')\n",
    "        \n",
    "    \n",
    "    \n",
    "class Edge():\n",
    "    '''Represents an edge in a Neo4J database'''\n",
    "    requiredParameters={\n",
    "        'relation': 'relationship type',\n",
    "        'source': 'Node ',\n",
    "        'target': 'Node '}\n",
    "    optionalParams={\n",
    "        'startyear': 'Academic year start',\n",
    "        'endyear': 'Academic year to'\n",
    "        }\n",
    "    def __init__(self, factory, **kwargs):\n",
    "        '''Edge between two nodes. Must specify source, target and relationship type.'''\n",
    "        for kw in self.requiredParameters:\n",
    "            if kw not in kwargs:\n",
    "                raise InsufficientParameterSpecException(f\"{kw} not specified in Edge constructor\")\n",
    "            self.params[kw]=kwargs[kw]\n",
    "    def set_expiry(self, expirydate):\n",
    "        '''\n",
    "        Sets the 'to' parameter with an academic year for when this relation was withdrawn. \n",
    "        The date specified will be the last academic year for which it was valid.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        expirydate : Academic year\n",
    "            Academic year is an academic year object or text description of it.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        self.params['to'] =AcademicYear(expirydate)\n",
    "    \n",
    "    def isCurrent(self, year):\n",
    "        '''\n",
    "        If the edge has a start/end year, check if the given year is the period in which the relation is valid\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        Boolean: True if the eyar is between start and end (inclusive), after start if no end, or True if no start\n",
    "\n",
    "        '''\n",
    "        \n",
    "        if self.params.get('startyear'):\n",
    "            if AcademicYear(year)< AcademicYear(self.params.get('startyear')):\n",
    "                return False\n",
    "            if self.params.get('endyear'):\n",
    "                if AcademicYear(year)> AcademicYear(self.params.get('endyear')):\n",
    "                    return False\n",
    "        return True\n",
    "    def getSource(self):\n",
    "        '''\n",
    "        Returns a Python object representing the source node\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        \n",
    "class CurriculumFactory():\n",
    "    \n",
    "    def __init__(self, db, dbname='curriculum'):\n",
    "        self.db = db\n",
    "        self.dbname = dbname\n",
    "        self.programmecache = {}\n",
    "        self.modulecache = {}\n",
    "        self.activitycache = {}\n",
    "        self.personcache = {}\n",
    "        \n",
    "    def get_connection(self):\n",
    "        return self.db\n",
    "    \n",
    "    @lru_cache    \n",
    "    def get_element_by_ID(self, elementID):\n",
    "        '''\n",
    "        Retrieves an element by ID and creates the appropriate node for it, if it is a node.\n",
    "        \n",
    "        Parameters\n",
    "        ----------\n",
    "        elementID : Internal neo4j element ID\n",
    "            text string uniquely identifying the object in the database.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        Object of correct class, if available or None if not found.\n",
    "\n",
    "        '''\n",
    "        erecords, _, _ = self.db.execute_query(\n",
    "            \"MATCH (a ) where elementId(a)=$id\"\n",
    "            \"RETURN a\", id=elementID,\n",
    "             database_=self.dbname,\n",
    "        )\n",
    "        if not erecords :\n",
    "            return\n",
    "        nodeclass = list(erecords[0].items()[0][1].labels())[0]\n",
    "        params = dict(erecords[0].items()[0][1])\n",
    "        params['elementID']=elementID\n",
    "        if nodeclass in globals():\n",
    "            return globals()[nodeclass](**params)\n",
    "        \n",
    "      \n",
    "    def get_or_create_Element(self, ElementName,**kwargs):\n",
    "        '''\n",
    "        Generic get and create method. This requires all elements to have a unique id under the property code\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        ElementName : Name of element to create, eg Programme\n",
    "            This will directly map to the class name.\n",
    "        **kwargs : Parameters to use. These will be checked against every class.\n",
    "            \n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        Element of type ElementName on success\n",
    "\n",
    "        '''\n",
    "        \n",
    "        elem= None\n",
    "        \n",
    "        if ElementName not in globals():\n",
    "            return\n",
    "        ElementClass=globals()[ElementName]\n",
    "        if kwargs.get('elementID',None):\n",
    "            elem = self.get_element_by_id(kwargs['elementID'])\n",
    "            return elem\n",
    "        if kwargs.get('code'):\n",
    "            try:\n",
    "                result,_,_=self.db.execute_query(f'MATCH (p:{ElementName} {{code: $name}} ) return p', name=kwargs['code'], database_=self.dbname)\n",
    "                if result:\n",
    "                    params=dict(result[0].items()[0][1])\n",
    "                    params['elementID']=result[0]['p'].element_id\n",
    "                    elem = ElementClass(self, **params)\n",
    "                    return elem\n",
    "            except Exception as e:\n",
    "                e.add_note(str(params))\n",
    "                raise e\n",
    "                \n",
    "        for p in ElementClass.requiredParams:\n",
    "            if p not in kwargs:\n",
    "                raise InsufficientParameterSpecException(f'Not enough parameters specified for {ElementName}: missing{p}')\n",
    "        \n",
    "        elem = ElementClass(self, **kwargs)\n",
    "        return elem\n",
    "\n",
    "\n",
    "    def get_all_elements(self, ElementName, **params):\n",
    "        '''\n",
    "        Returns all elements of type ElementName\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        ElementName : name of an element\n",
    "            This can be any element type for which a model exists.\n",
    "        **params : Keyword parameters to use in the search\n",
    "            Exact matches only at present.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        List of elements\n",
    "\n",
    "        '''\n",
    "        \n",
    "        elements = []\n",
    "        if ElementName not in globals():\n",
    "            return elements\n",
    "        paramdetails =[]\n",
    "        paramtext=''\n",
    "        for p in params:\n",
    "            paramdetails.append(f'{p}: ${p}')\n",
    "        if paramdetails:\n",
    "            paramtext= f'{ {\",\".join(paramdetails)} }'\n",
    "        erecords, _, _ = self.db.execute_query(\n",
    "            f\"MATCH (a :{ElementName} {paramtext})\",\n",
    "            params,\n",
    "             database_=self.dbname,\n",
    "        )\n",
    "        if not erecords :\n",
    "            return elements\n",
    "        for item in erecords:\n",
    "            elem=item.items()[0][1]\n",
    "            nodeclass = list(elem.labels())[0]\n",
    "            params = dict(elem)\n",
    "            params['elementID']=item.element_id\n",
    "            if nodeclass in globals():\n",
    "                elements.append( globals()[nodeclass](self, **params))\n",
    "        return elements  \n",
    "    def getElementsForElement(self, element, target, max_steps=2,relation=None):\n",
    "        '''\n",
    "        Retreive all elements of type Target linked to element,optionally by relation (or all relations) \n",
    "        at a maximum distance of max_steps (default 2)\n",
    "        \n",
    "        Parameters\n",
    "        ----------\n",
    "        element : TYPE\n",
    "            DESCRIPTION.\n",
    "        target : TYPE\n",
    "            DESCRIPTION.\n",
    "        max_steps: integer\n",
    "            Maximum link number to explore (default 2)\n",
    "        relation: text\n",
    "            Limit relations to those of type relation\n",
    "        Returns\n",
    "        -------\n",
    "        List of Node type objects.\n",
    "        '''\n",
    "        if target not in globals():\n",
    "            return []\n",
    "        cypher = f\"MATCH (e) --{{1,{max_steps}}} (thing:$target) where elementID(e)=$id RETURN DISTINCT thing\"\n",
    "        result,_,_=self.db.execute_query(cypher, target=target, id=element.element_id,database_=self.dbname)\n",
    "        elems = []\n",
    "        if result:\n",
    "            for elem in result:\n",
    "                item = elem.items()[0][1]\n",
    "                params =dict(item)\n",
    "                params['elementID'] = item.element_id\n",
    "                elems.append(globals()[target](self, **params))\n",
    "        return elems\n",
    "         \n",
    "class Programme(Node):\n",
    "    \n",
    "    DRAFT = 0\n",
    "    CURRENT = 1\n",
    "    ARCHIVED = 2\n",
    "    WITHDRAWN =3\n",
    "    ntype='Programme'\n",
    "    requiredParams = {\n",
    "        'code': 'Programme code',\n",
    "        'name': 'Programme name'\n",
    "        }\n",
    "    optionalParams= {\n",
    "        'startyear':'First academic year',\n",
    "        'endyear': 'Last academic year',\n",
    "        'school':'School which manages the Programme'}\n",
    "    def __init__(self, factory, **params ):\n",
    "        '''\n",
    "        Create a Programme instance.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        factory - database connection and cache for Programme\n",
    "        **params : \n",
    "            Create or replace table Programme (\n",
    "        '''\n",
    "        super().__init__(factory,**params)\n",
    "        \n",
    "        self.modules = {}\n",
    "        self.ILO = []\n",
    "        if self.element_id:\n",
    "            self.loadmodules()\n",
    "            self.loadILO()\n",
    "        \n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    def loadmodules(self):\n",
    "        '''\n",
    "        Loads module links from the database to the modules list. Does not create now entries in the database.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        self.modules={}\n",
    "        cypher = \"MATCH (p:Programme ) -[a]-(b:Module) WHERE elementID(p) = $id RETURN a,b\"\n",
    "        records,_,_ =self.factory.db.execute_query(cypher, id=self.element_id, database_=self.factory.dbname)\n",
    "        for t in records:\n",
    "            relation = t.items()[0][1]\n",
    "            target = t.items()[1][1]\n",
    "            if target.data()['code'] not in self.modules:\n",
    "                self.modules[target.data()['code']]=[]\n",
    "            self.modules[target.data()['code']].append({'relation': {'type': relation.type, 'params':dict(relation)}, \n",
    "                                                        'target':{ 'moduleID': target.element_id,'params':dict(target)}})\n",
    "            \n",
    "        \n",
    "\n",
    "        \n",
    "    def loadILO(self):\n",
    "        '''\n",
    "        Loads ILO links from the database. Does not create new links or ILOs.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        self.ILO = {}\n",
    "        cypher = \"MATCH (p:Programme ) -[a]-(b:ProgrammeILO) WHERE elementID(p) = $id RETURN a,b\"\n",
    "        records,_,_ =self.factory.db.execute_query(cypher, id=self.element_id, database_=self.factory.dbname)\n",
    "        for t in records:\n",
    "            relation, target = t.items()[0:2]\n",
    "            \n",
    "            self.ILO[target.element_id]= (dict(target), dict(relation))\n",
    "        \n",
    " \n",
    "        \n",
    "    def withdraw(self, year):\n",
    "        '''\n",
    "        Set the Programme to WITHDRAWN. Only possible if there are no future versions.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        if AcademicYear(year) and AcademicYear(year) > AcademicYear(self.params['startyear']):\n",
    "            self.params['endyear']=year\n",
    "            cypher = 'MATCH (p:Programme) where elementID(p)=$id SET p.endyear=$year'\n",
    "            records,_,_ =self.factory.db.execute_query(cypher, id=self.element_id,year=year, database_=self.factory.dbname)\n",
    "        \n",
    "\n",
    "    def map_module(self, module, optional=0, year=None, remove=False):\n",
    "        '''\n",
    "        Add a module to the Programme, updating if exisitng  already appended.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        module : Module object\n",
    "            DESCRIPTION.\n",
    "        optional: Boolean\n",
    "        remove: Academic Year for last instance\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        relations=('IS_CORE','IS_ELECTIVE')\n",
    "        relation= relations[int(bool(optional))]\n",
    "        cypher1 = 'MATCH (p:Programme) -[b:{relation}]-(m:Module) where elementID(p)=$pid and elementID(m)=$mid SET b.endyear=$year'\n",
    "        \n",
    "        if module.params['code'] in self.modules:\n",
    "            if remove:\n",
    "                \n",
    "                records,_,_ =self.factory.db.execute_query(cypher1.format(relation=relation), \n",
    "                                                           pid=self.element_id,year=year, \n",
    "                                                           mid=module.element_id,\n",
    "                                                           database_=self.factory.dbname)\n",
    "                for r in self.modules[module.params['code']]:\n",
    "                    if r['relation']['type']==relation:\n",
    "                        r['relation']['params']['endyear']=year\n",
    "            else:\n",
    "                for r in self.modules[module.params['code']]:\n",
    "                    if r['relation']['type']!=relation:\n",
    "                        r['relation']['params']['endyear']=year\n",
    "                        records,_,_ =self.factory.db.execute_query(cypher1.format(relation=relations[(int(bool(optional))+1)%2]), \n",
    "                                                                   pid=self.element_id,year=year, \n",
    "                                                                   mid=module.element_id,\n",
    "                                                                   database_=self.factory.dbname)\n",
    "         \n",
    "        if not remove:\n",
    "            if module.params['code'] not in self.modules:\n",
    "                 self.modules[module.params['code']]=[]\n",
    "            cypher2 =f'MERGE (p:Programme) <-[{relation} {{startyear:$year }}] - (m:Module) where elementID(p)=$pid and elementID(m)=$mid'\n",
    "            records,_,_ =self.factory.db.execute_query(cypher2, \n",
    "                                                        pid=self.element_id,year=year, \n",
    "                                                        mid=module.element_id,\n",
    "                                                        database_=self.factory.dbname)\n",
    "        self.loadmodules()\n",
    "           \n",
    "                                           \n",
    "               \n",
    "           \n",
    "           \n",
    "    def map_ilo(self, ilo, year, remove=False):\n",
    "        '''\n",
    "        Associates or updates a Programme ILO mapping to a Programme.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        ilo : ProgrammeILO \n",
    "            Programme ILO object\n",
    "        year : text describing the academic year\n",
    "            Must be either the start year, or for removing an ILO the last year for which it will be relevant.\n",
    "        remove : Boolean, optional\n",
    "            Flag to say whether the ILO should be dissociated from teh Programme. The default is False.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        \n",
    "        if not AcademicYear(year):\n",
    "            raise UnparseableYearException(f'Cannot parse year value {year}')\n",
    "        if ilo.element_id in self.ILO:\n",
    "            if remove:\n",
    "                self.ILO[ilo.element_id]['endyear']=year\n",
    "                cypher=\"MATCH (p: Programme) -[b]- (i:ProgrammeILO) WHERE elementID(p) =$pid AND elementID(i) =$iid SET b.endyear=$year\"\n",
    "                records,_,_ = self.factory.db.execute_query(cypher,pid=self.element_id, iid=ilo.element_id, year=year, database_=self.factory.dbname)\n",
    "            return\n",
    "        if remove:\n",
    "            return\n",
    "        cypher = \"MERGE (p:Programme) -[b:HAS_ILO {startyear:$year}]->(i:ProgrammeILO) where elementID(p)=$pid AND elementID(i) = $iid RETURN b\"\n",
    "        records,_,_ = self.factory.db.execute_query(cypher,pid=self.element_id, iid=ilo.element_id, year=year, database_=self.factory.dbname)\n",
    "        if records:\n",
    "            relation=records[0].items()[0][1]\n",
    "            self.ILO[ilo.element_id]=(dict(ilo),dict(relation))\n",
    "\n",
    "\n",
    "    \n",
    "class Module(Node):\n",
    "    \n",
    "    DRAFT = 0\n",
    "    CURRENT = 1\n",
    "    ARCHIVED = 2\n",
    "    WITHDRAWN = 3\n",
    "    PREREQUISITE = 1\n",
    "    COREQUISITE = 0\n",
    "    ANTIREQUISITE = 2\n",
    "    ntype='Module'\n",
    "    requiredParams = {\n",
    "        'code': 'Module code',\n",
    "        'name': 'Module name',\n",
    "        'credits': 'Credit weight',\n",
    "        'scqflevel': 'SCQF level',\n",
    "        'shelevel': 'SHE level',\n",
    "        'semester': '1, 2 or both'\n",
    "        }\n",
    "    optionalParams= {\n",
    "        'block': 'Teaching block: C or D',\n",
    "        'startyear':'First academic year',\n",
    "        'endyear': 'Last academic year'\n",
    "        }\n",
    "    \n",
    "    def __init__(self, factory,  **params ):\n",
    "        '''\n",
    "        Create a Programme instance.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        factory - database connection and cache for Module\n",
    "        **params : \n",
    "                    '''\n",
    "        super().__init__(factory,**params)\n",
    "        \n",
    "        self.ILO={}\n",
    "        self.Activities={}\n",
    "        self.Assessments = {}\n",
    "        if self.element_id:\n",
    "            self.loadILO()\n",
    "            self.loadConstraints()\n",
    "            self.loadActivities()\n",
    "            self.loadAssessments()\n",
    "            \n",
    "    def loadActivities(self):\n",
    "        '''\n",
    "        Loads module links from the database to the modules list. Does not create now entries in the database.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        self.Activities=[]\n",
    "        cypher = \"MATCH (m:Module ) -[a]-(b:TeachingActivity) WHERE elementID(m) = $id RETURN a,b\"\n",
    "        records,_,_ =self.factory.db.execute_query(cypher, id=self.element_id, database_=self.factory.dbname)\n",
    "        for t in records:\n",
    "            relation = t.items()[0][1]\n",
    "            target = t.items()[1][1]\n",
    "            self.Activities.append({'relation': {'type': relation.type, 'params':dict(relation)}, \n",
    "                                                        'target':{ 'TeachingActivityID': target.element_id,'params':dict(target)}})\n",
    "            \n",
    "        \n",
    "        \n",
    "        \n",
    "    def loadILO(self):\n",
    "        '''\n",
    "        Loads ILO links from the database. Does not create new links or ILOs.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        self.ILO = {}\n",
    "        cypher = \"MATCH (m:Module ) -[a]-(b:ModuleILO) WHERE elementID(m) = $id RETURN a,b\"\n",
    "        records,_,_ =self.factory.db.execute_query(cypher, id=self.element_id, database_=self.factory.dbname)\n",
    "        for t in records:\n",
    "            relation, target = t.items()[0:2]\n",
    "            self.ILO[target.element_id]= (dict(target),dict(relation))\n",
    "        \n",
    "\n",
    "    def loadConstraints(self):\n",
    "        '''\n",
    "        Loads in but does not create constraint\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        ''' \n",
    "        self.requisites = {}\n",
    "        cypher = \"MATCH (m:Module ) <-[a]- (b:Module) WHERE elementID(m) = $id RETURN a,b\"\n",
    "        records,_,_ =self.factory.db.execute_query(cypher, id=self.element_id, database_=self.factory.dbname)\n",
    "        for t in records:\n",
    "            relation, target = t.items()[0:2]\n",
    "            relationdata = dict(relation)\n",
    "            relationdata['element_id']=relation.element_id\n",
    "            relationdata['type']=relation.type\n",
    "            targetdata = dict(target)\n",
    "            targetdata['element_id']=target.element_id\n",
    "            self.requisites.append({'constraint':relationdata, 'module': targetdata})\n",
    "    def loadAssessments(self):\n",
    "        '''Retrieve any linked assessments from the database\n",
    "        # TODO\n",
    "        '''\n",
    "        #TODO\n",
    "        \n",
    "    def setConstraint(self, constraint, module, year, remove=False):\n",
    "        '''\n",
    "        Constraint is an integer corresponding to the constraint type. \n",
    "        module is teh module acting as a constraint.\n",
    "        remove is boolean actig as to whether a constraint is added/updated (false) or removed (true)\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        constraintlist : TYPE\n",
    "            DESCRIPTION.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        #TODO\n",
    "            \n",
    "        \n",
    "        \n",
    "        \n",
    "    def map_ilo(self, ilo, year, remove=False):\n",
    "        '''\n",
    "        Associates or updates a Module ILO mapping to a Module.\n",
    "    \n",
    "        Parameters\n",
    "        ----------\n",
    "        ilo : ModuleILO \n",
    "            Module ILO object\n",
    "        year : text describing the academic year\n",
    "            Must be either the start year, or for removing an ILO the last year for which it will be relevant.\n",
    "        remove : Boolean, optional\n",
    "            Flag to say whether the ILO should be dissociated from the Module. The default is False.\n",
    "    \n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "    \n",
    "        '''\n",
    "        \n",
    "        if not AcademicYear(year):\n",
    "            raise UnparseableYearException(f'Cannot parse year value {year}')\n",
    "        if ilo.element_id in self.ILO:\n",
    "            if remove:\n",
    "                self.ILO[ilo.element_id]['endyear']=year\n",
    "                cypher=\"MATCH (m: Module) -[b]- (i:ModuleILO) WHERE elementID(m) =$mid AND elementID(i) =$iid SET b.endyear=$year\"\n",
    "                records,_,_ = self.factory.db.execute_query(cypher,mid=self.element_id, iid=ilo.element_id, year=year, database_=self.factory.dbname)\n",
    "            return\n",
    "        if remove:\n",
    "            return\n",
    "        cypher = \"MERGE (m:Module) -[b:HAS_ILO {startyear:$year}]->(i:ModuleILO) where elementID(m)=$mid AND elementID(i) = $iid RETURN b\"\n",
    "        records,_,_ = self.factory.db.execute_query(cypher,mid=self.element_id, iid=ilo.element_id, year=year, database_=self.factory.dbname)\n",
    "        if records:\n",
    "            relation=records[0].items()[0][1]\n",
    "            self.ILO[ilo.element_id]=(dict(ilo),dict(relation))\n",
    "    \n",
    "    def map_activity( self, activity, relationtype='HAS_ACTIVITY'):\n",
    "        '''\n",
    "        Adds an activity to a module\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        activity : TeachingActivity\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        if activity.element_id not in [x['target']['TeachingActivityID'] for x in self.Activities]:\n",
    "            cypher = f\"MERGE (m:Module ) -[a:{relationtype}]-(b:TeachingActivity) WHERE elementID(m) = $id AND elementID(b)=$tid RETURN a,b\"\n",
    "            records,_,_ =self.factory.db.execute_query(cypher, id=self.element_id, tid=activity.element_id, database_=self.factory.dbname)\n",
    "            for t in records:\n",
    "                relation = t.items()[0][1]\n",
    "                target = t.items()[1][1]\n",
    "                self.Activities.append({'relation': {'type': relation.type, 'params':dict(relation)}, \n",
    "                                                            'target':{ 'TeachingActivityID': target.element_id,'params':dict(target)}})\n",
    "        \n",
    "\n",
    "        \n",
    "# class TeachingActivityType(Node): \n",
    "    \n",
    "#     ntype='TeachingActivityType'\n",
    "#     requiredParams = {\n",
    "#         'code': 'Module code',\n",
    "#         'name': 'Module name',\n",
    "#         'credits': 'Credit weight',\n",
    "#         'level': 'SCQF level'\n",
    "#         }\n",
    "#     optionalParams= {\n",
    "#         'startyear':'First academic year',\n",
    "#         'endyear': 'Last academic year',\n",
    "#         'school':'School which manages the Programme'}\n",
    "\n",
    "    \n",
    "#     def __init__(self, factory, **kwargs):\n",
    "#         self.factory=factory\n",
    "#         self.name = kwargs.get(\"name\")\n",
    "#         self.definition = kwargs.get(\"definition\")\n",
    "#         self.id = kwargs.get(\"id\")\n",
    "#         if self.id is None:\n",
    "#             self.create()\n",
    "            \n",
    "        \n",
    "#     def create(self):\n",
    "#         '''\n",
    "#         Creates a new database entry for the TeachingActivityType entity\n",
    "\n",
    "#         Returns\n",
    "#         -------\n",
    "#         None.\n",
    "\n",
    "#         '''\n",
    "#         insertsql = \"Insert into TeachingActivityType (name, definition) VALUES (%s,%s)\"\n",
    "#         cursor = self.factory.db.cursor()\n",
    "#         cursor.excecute(insertsql, (self.name, self.definition))\n",
    "#         self.id = cursor.lastrowid\n",
    "\n",
    "\n",
    "\n",
    "class TeachingActivity (Node):\n",
    "    ntype='TeachingActivity'\n",
    "    requiredParams = {\n",
    "        'year': 'AcademicYear',\n",
    "        'type': 'Activity type'\n",
    "        }\n",
    "    optionalParams= {\n",
    "        'duration':'length of activity'\n",
    "        }\n",
    "\n",
    "    def __init__(self, factory, **kwargs):\n",
    "        #description, duration, TAtype, version,previous_TA, moduleID, sequence\n",
    "        super().__init__(factory, **kwargs)\n",
    "        self.loadILO()\n",
    "        \n",
    "    def loadILO(self):\n",
    "        '''\n",
    "        Loads ILO links from the database. Does not create new links or ILOs.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        self.ILO = {}\n",
    "        cypher = \"MATCH (t:TeachingActivity ) -[a]-(b:ActivityILO) WHERE elementID(t) = $id RETURN a,b\"\n",
    "        records,_,_ =self.factory.db.execute_query(cypher, id=self.element_id, database_=self.factory.dbname)\n",
    "        for t in records:\n",
    "            relation, target = t.items()[0:2]\n",
    "            self.ILO[target.element_id]= (dict(target),dict(relation))\n",
    "    def map_ilo(self, ilo):\n",
    "        '''\n",
    "        Associates or updates a Module ILO mapping to a Module.\n",
    "    \n",
    "        Parameters\n",
    "        ----------\n",
    "        ilo : ModuleILO \n",
    "            Module ILO object\n",
    "        year : text describing the academic year\n",
    "            Must be either the start year, or for removing an ILO the last year for which it will be relevant.\n",
    "        remove : Boolean, optional\n",
    "            Flag to say whether the ILO should be dissociated from the Module. The default is False.\n",
    "    \n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "    \n",
    "        '''\n",
    "        \n",
    "        if ilo.element_id in self.ILO:\n",
    "            return\n",
    "        cypher = \"MERGE (t:TeachingActivity) -[b:HAS_ILO ]->(i:ActivityILO) where elementID(t)=$tid AND elementID(i) = $iid RETURN b\"\n",
    "        records,_,_ = self.factory.db.execute_query(cypher,tid=self.element_id, iid=ilo.element_id, database_=self.factory.dbname)\n",
    "        if records:\n",
    "            relation=records[0].items()[0][1]\n",
    "            self.ILO[ilo.element_id]=(dict(ilo),dict(relation))\n",
    "            \n",
    "\n",
    "    def assign_to_module(self, module):\n",
    "        '''\n",
    "        Add an activity to a module\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        module : A TModule object to attach the activity to.\n",
    "            DESCRIPTION.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        module.map_activity(self)\n",
    "        self.module=module\n",
    "        \n",
    "\n",
    "        \n",
    "        \n",
    "        \n",
    "class ILO(Node):\n",
    "    ntype='ILO'\n",
    "    ILO_KNOWLEDGE = 1\n",
    "    ILO_UNDERSTANDING=2\n",
    "    ILO_SKILL=3\n",
    "    ILO_ATTITUDE=4\n",
    "    requiredParams = {\n",
    "        }\n",
    "    optionalParams= {\n",
    "        'SOLO': 'Solo level of the ILO'\n",
    "        }\n",
    "    def __init__(self, factory,**kwparams):\n",
    "        super().__init__(factory,**kwparams)\n",
    "    \n",
    "    \n",
    "\n",
    "class ActivityILO(ILO):\n",
    "\n",
    "    bloomterms = ('None','Remember',\t'Understand',\t'Apply',\t'Analyze',\t'Evaluate',\t'Create')\n",
    "    bloommap ={\n",
    "        }\n",
    "    categories = ('Knowledge','Understanding', 'Skill', 'Attitude')\n",
    "\n",
    "    def __init__(self, factory, **kwargs):\n",
    "        super().__init__(factory, **kwargs)\n",
    "        self.milo=None\n",
    "        self._get_milo()\n",
    "      \n",
    "    def _get_milo(self):\n",
    "        '''\n",
    "        Retrieves and assigns any mapped MILO to this AILO. \n",
    "        Raises an exception if more than one MILO is mapped in the DB\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        relation object between AILO and AILO\n",
    "\n",
    "        '''\n",
    "       \n",
    "        cypher = \"MATCH (a:ActivityILO) -[b:MAPS_TO] - (m:ModuleILO) WHERE elementID(a)=$id RETURN a,b,m\"\n",
    "        records,_,_ = self.factory.db.execute_query(cypher,id=self.element_id, \n",
    "                                                    database_=self.factory.dbname)\n",
    "        if records:\n",
    "            if len(records)>1:\n",
    "                raise Exception('Maps to multiple Module ILOs')\n",
    "            \n",
    "            target=records[0].items()[2][1]\n",
    "            self.milo = target    \n",
    "    \n",
    "    def getActivitiesForILO(self):\n",
    "        '''\n",
    "        Retrieve all Activities with this ILO\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        List of TeachingActivity\n",
    "\n",
    "        '''\n",
    "        activities = self.factory.getElementsForElement(self,'TeachingActivity',max_steps=1)\n",
    "        return activities\n",
    "\n",
    "    def supports(self, milo=None):\n",
    "        '''\n",
    "        Maps (if milo given) and returns the moduleILO, if present,\n",
    "        to which this activity ILO maps\n",
    "        \n",
    "        \n",
    "        Parameters\n",
    "        ----------\n",
    "        milo : ModuleILO, optional\n",
    "            ModuleILO object which should be mapped to this activity ILO. \n",
    "            The default is None.\n",
    "        \n",
    "        Returns\n",
    "        -------\n",
    "        ModuleILO if assigned, None if not assigned\n",
    "        \n",
    "        '''\n",
    "        if self.milo and milo:\n",
    "            if milo != self.milo:\n",
    "                #delete existing relation and replace.\n",
    "                cypherd = \"MATCH (a:ActivityILO) -[b:MAPS_TO] - (m:ModuleILO) WHERE elementID(a)=$id DELETE b\"\n",
    "                records,_,_ = self.factory.db.execute_query(cypherd,id=self.element_id, \n",
    "                                                            database_=self.factory.dbname)\n",
    "            else:\n",
    "                return self.milo\n",
    "        if milo:\n",
    "            cyphera = \"MERGE (a:ActivityILO) -[MAPS_TO]-> (m:ModuleILO) WHERE elementID(a)=$id AND elementID(m) =$mid\"\n",
    "            records,_,_ = self.factory.db.execute_query(cyphera,\n",
    "                                                        id=self.element_id, mid=milo.element_id,\n",
    "                                                        database_=self.factory.dbname)\n",
    "            self.milo=milo\n",
    "        return self.milo\n",
    "\n",
    "    \n",
    "class ModuleILO(ILO):\n",
    "     \n",
    "    def __init__(self, factory, **kwargs):\n",
    "        super().__init__(factory, **kwargs)\n",
    "        self.pilo=None\n",
    "        self._get_pilo()\n",
    "\n",
    "    def _get_pilo(self):\n",
    "        '''\n",
    "        Retrieves and assigns any mapped PILO to this MILO. \n",
    "        Raises an exception if more than one PILO is mapped in the DB\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        relation object between MILO and PILO\n",
    "\n",
    "        '''\n",
    "        \n",
    "        cypher = \"MATCH (m:ModuleILO) -[a:MAPS_TO] - (p:ProgrammeILO) WHERE elementID(m)=$id RETURN m,a,p\"\n",
    "        records,_,_ = self.factory.db.execute_query(cypher,id=self.element_id, \n",
    "                                                    database_=self.factory.dbname)\n",
    "        if records:\n",
    "            if len(records)>1:\n",
    "                raise Exception('Maps to multiple Programme ILOs')\n",
    "            \n",
    "            target=records[0].items()[2][1]\n",
    "            self.pilo = target\n",
    "\n",
    "    def getModulesForILO(self):\n",
    "        '''\n",
    "        Retrieve all Activities with this ILO\n",
    "         \n",
    "        Returns\n",
    "        -------\n",
    "        List of Modules\n",
    "         \n",
    "        '''\n",
    "        modules = self.factory.getElementsForElement(self,'Module',max_steps=1)\n",
    "        return modules\n",
    "            \n",
    "    def supports(self, pilo=None):\n",
    "        '''\n",
    "        Maps (if pilo given) and returns the progammeILO, if present to which this module ILO maps\n",
    "        \n",
    "        \n",
    "        Parameters\n",
    "        ----------\n",
    "        pilo : ProgrammeILO, optional\n",
    "            ProgrammeILO object which should be mapped to this module ILO. \n",
    "            The default is None.\n",
    "        \n",
    "        Returns\n",
    "        -------\n",
    "        ProgrammeILO if assigned, None if not assigned\n",
    "        \n",
    "        '''\n",
    "        if self.pilo and pilo:\n",
    "            if pilo != self.pilo:\n",
    "                #delete existing relation and replace.\n",
    "                cypherd = \"MATCH (m:ModuleILO) -[a:MAPS_TO] - (p:ProgrammeILO) WHERE elementID(m)=$id DELETE a\"\n",
    "                records,_,_ = self.factory.db.execute_query(cypherd,id=self.element_id, \n",
    "                                                            database_=self.factory.dbname)\n",
    "            else:\n",
    "                return self.pilo\n",
    "        if pilo:\n",
    "            cyphera = \"MERGE (m:ModuleILO) -[MAPS_TO]-> (p:ProgrammeILO) WHERE elementID(m)=$id AND elementID(p) =$pid\"\n",
    "            records,_,_ = self.factory.db.execute_query(cyphera,\n",
    "                                                        id=self.element_id, pid=pilo.element_id,\n",
    "                                                        database_=self.factory.dbname)\n",
    "            self.pilo=pilo\n",
    "        return self.pilo\n",
    "\n",
    "    \n",
    "class ProgrammeILO(ILO):\n",
    "   \n",
    "     \n",
    "    def __init__(self, factory, **kwargs):\n",
    "        super().__init__(factory, **kwargs)\n",
    "        self.pilo=None\n",
    "\n",
    "    def getProgrammesForILO(self):\n",
    "        '''\n",
    "        Retrieve all Activities with this ILO\n",
    "         \n",
    "        Returns\n",
    "        -------\n",
    "        List of Modules\n",
    "         \n",
    "        '''\n",
    "        programmes = self.factory.getElementsForElement(self,'Programme',max_steps=1)\n",
    "        return programmes\n",
    "            \n",
    "     \n",
    "    \n",
    "class AssessmentType():\n",
    "    '''\n",
    "    CREATE or REPLACE TABLE AssessmentType (\n",
    "    ID integer not null primary key auto_increment,\n",
    "    name text not null,\n",
    "    definition text not null\n",
    "    );\n",
    "    '''\n",
    "\n",
    "    def __init__(self, factory, **kwargs):\n",
    "        self.factory = factory\n",
    "        self.name = kwargs.get(\"name\")\n",
    "        self.definition = kwargs.get(\"definition\")\n",
    "        self.id = kwargs.get(\"id\")\n",
    "        if self.id is None:\n",
    "            self.create()\n",
    "            \n",
    "        \n",
    "    def create(self):\n",
    "        '''\n",
    "        Creates a new database entry for the AssessmentType entity\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        None.\n",
    "\n",
    "        '''\n",
    "        insertsql = \"Insert into AssessmentType (name, definition) VALUES (%s,%s)\"\n",
    "        cursor = self.factory.db.cursor()\n",
    "        cursor.excecute(insertsql, (self.name,self.definition))\n",
    "        self.id = cursor.lastrowid   \n",
    "\n",
    "class Assessment(TeachingActivity):\n",
    "    ntype='Assessment'\n",
    "    requiredParams = {\n",
    "        'year': 'AcademicYear',\n",
    "        'type': 'Activity type',\n",
    "        'weighting': 'percentage weighting',\n",
    "        'code': 'SEQ number'\n",
    "        }\n",
    "    optionalParams= {\n",
    "        }\n",
    "\n",
    "    def __init__(self, factory,**kwargs):\n",
    "        super().__init__(factory, **kwargs)\n",
    "        \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "f512dcc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "factory = CurriculumFactory(driver, 'curriculumdb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "9de8ea4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "bs32031=factory.get_or_create_Element('Module', **newmodparams)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "06125e43",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Module: [code: BS32031, name: Bioinformatics Research Skills 1, credits: 15, scqflevel: 3, shelevel: 9, semester: 2] ID: 4:4f947b4e-c5e2-45c5-b439-76aa1c96d891:0"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bs32031"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "9a1865e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Module', 'Node', 'object']"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[x.__name__ for x in globals()['Module'].__mro__]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d61f210e",
   "metadata": {},
   "outputs": [],
   "source": [
    "doc = docx.Document(\"BS32008 Plant Science Specification QASv3.docx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb1931c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "for p in doc.paragraphs:\n",
    "    print(p.text)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ac2a095",
   "metadata": {},
   "outputs": [],
   "source": [
    "doc.tables[0].row_cells(0)[1].text\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "076d0955",
   "metadata": {},
   "outputs": [],
   "source": [
    "tcount=0\n",
    "structure ={}\n",
    "for t in doc.tables:\n",
    "    print(f'Table {tcount} ({len(t.rows)} rows)')\n",
    "    tcount+=1\n",
    "    for r in range(len(t.rows)):\n",
    "        print(f'row {r} [{t.row_cells(r)[0].text}] [{t.row_cells(r)[1].text}])')\n",
    "        structure[t.row_cells(r)[0].text.split()[0]] =t.row_cells(r)[1].text\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8bf731a",
   "metadata": {},
   "outputs": [],
   "source": [
    "structure.keys()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
